{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read(\"lyon_transcriptomics_scanpy_normalised.h5\")\n",
    "adata.obs[\"Metadata_treatments\"] = adata.obs[\"Metadata_treatments\"].astype(str)\n",
    "adata.obs[\"Metadata_target\"] = adata.obs[\"Metadata_target\"].astype(str)\n",
    "adata.obs[\"Metadata_Mode_of_Action\"] = adata.obs[\"Metadata_Mode_of_Action\"].astype(str)\n",
    "adata.obs.loc[adata.obs[\"Metadata_treatments\"] == \"DMSO\", \"Metadata_target\"] = \"DMSO\"\n",
    "adata.obs.loc[adata.obs[\"Metadata_treatments\"] == \"DMSO\", \"Metadata_Mode_of_Action\"] = \"DMSO\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Metadata_Mode_of_Action\n",
       "lipid SBI           245\n",
       "respiration         178\n",
       "lipid other         129\n",
       "GPI                 102\n",
       "cytoskeleton         94\n",
       "amino acid           72\n",
       "prot metabolism      58\n",
       "proteins             57\n",
       "DMSO                 56\n",
       "DHODH                56\n",
       "cell wal             44\n",
       "HOG                  43\n",
       "Cell wall            32\n",
       "Multi                30\n",
       "HDAC                 28\n",
       "ATP proton pump      25\n",
       "GPI                  25\n",
       "IMPDH                25\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# rename_dict = {\n",
    "#     \"cell wal\" : 'Cell wall',\n",
    "#     \"lipid other\": \"lipid\",\n",
    "#     \"lipid SBI\": \"lipid\",\n",
    "# }\n",
    "# adata.obs[\"Metadata_Mode_of_Action\"].replace(rename_dict, inplace=True)\n",
    "# adata.obs[\"Metadata_Mode_of_Action\"] = adata.obs[\"Metadata_Mode_of_Action\"].str.capitalize()\n",
    "\n",
    "# adata = adata[adata.obs[\"Metadata_Mode_of_Action\"].isin(adata.obs[\"Metadata_Mode_of_Action\"].value_counts()[adata.obs[\"Metadata_Mode_of_Action\"].value_counts() >= 10].index)]\n",
    "adata.obs[\"Metadata_Mode_of_Action\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "gene_mapping = pickle.load(open(\"gene_mapping_botrytis.pkl\", \"rb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "pickle.dump(gene_mapping, open(\"gene_mapping_botrytis.pkl\", \"wb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var[\"Gene Name\"] = [gene_mapping[x] if x in gene_mapping else x for x in adata.var.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var[\"Gene Name\"] = adata.var[\"Gene Name\"].astype(str)\n",
    "# Check for duplicate gene names and add a suffix if necessary\n",
    "gene_name_counts = adata.var[\"Gene Name\"].value_counts()\n",
    "duplicate_gene_names = gene_name_counts[gene_name_counts > 1].index\n",
    "\n",
    "for gene_name in duplicate_gene_names:\n",
    "    indices = adata.var[adata.var[\"Gene Name\"] == gene_name].index\n",
    "    for i, idx in enumerate(indices):\n",
    "        adata.var.at[idx, \"Gene Name\"] = f\"{gene_name}_{i + 1}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['DMSO', 'lipid other ', 'lipid SBI ', 'prot metabolism ',\n",
       "       'respiration ', 'IMPDH', 'cell wal ', 'Cell wall ',\n",
       "       'cytoskeleton ', 'DHODH', 'GPI', 'HDAC', 'HOG', 'amino acid ',\n",
       "       'ATP proton pump ', 'GPI ', 'proteins ', 'Multi'], dtype=object)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[\"Metadata_Mode_of_Action\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gene_expression_per_group(subsample_adata, group, reference, genes2display=5):\n",
    "    sc.tl.rank_genes_groups(\n",
    "        subsample_adata,\n",
    "        group,\n",
    "        reference=reference,\n",
    "        # method=\"wilcoxon\",\n",
    "    )\n",
    "    # sc.pl.rank_genes_groups_dotplot(\n",
    "    #     subsample_adata,\n",
    "    #     n_genes = 5,\n",
    "    #     standard_scale='var',\n",
    "    #     gene_symbols = \"Gene Name\"\n",
    "    # )\n",
    "    # sc.pl.rank_genes_groups_dotplot(\n",
    "    #     adata,\n",
    "    #     n_genes = 5,\n",
    "    #     standard_scale='var',\n",
    "    #     gene_symbols = \"Summary functional categories\"\n",
    "    # )\n",
    "    subsample_adata = subsample_adata[subsample_adata.obs[group] != reference]\n",
    "\n",
    "    subsample_adata.obs[group] = subsample_adata.obs[group].astype(\"category\")\n",
    "    sc.tl.dendrogram(subsample_adata, groupby=group)\n",
    "\n",
    "    sc.pl.rank_genes_groups_dotplot(\n",
    "        subsample_adata,\n",
    "        n_genes=genes2display,\n",
    "        # standard_scale='var',\n",
    "        vmin=-3,\n",
    "        vmax=3,\n",
    "        cmap=\"bwr\",\n",
    "        values_to_plot=\"logfoldchanges\",\n",
    "        gene_symbols=\"Gene Name\",\n",
    "    )\n",
    "    # sc.pl.rank_genes_groups_heatmap(\n",
    "    #     subsample_adata,\n",
    "    #     n_genes=10,\n",
    "    #     use_raw=False,\n",
    "    #     swap_axes=True,\n",
    "    #     show_gene_labels=False,\n",
    "    #     vmin=-3,\n",
    "    #     vmax=3,\n",
    "    #     cmap=\"bwr\",\n",
    "    #     gene_symbols = \"Gene Name\"\n",
    "    # )\n",
    "    # sc.pl.rank_genes_groups_dotplot(\n",
    "    #     adata,\n",
    "    #     n_genes=5,  # Adjust to the number of genes you want to plot\n",
    "    #     #cmap = \"bwr\",\n",
    "    #     values_to_plot=\"logfoldchanges\",\n",
    "    #     gene_symbols = \"Description summary\"\n",
    "    # )\n",
    "\n",
    "    # sc.pl.rank_genes_groups(subsample_adata, n_genes=10, sharey=False, gene_symbols = \"Gene Name\")\n",
    "\n",
    "    # for member in subsample_adata.obs[group].unique():\n",
    "    #     if member == \"DMSO\":\n",
    "    #         continue\n",
    "    #     sc.pl.rank_genes_groups_violin(subsample_adata, groups=member, n_genes=5, gene_symbols = \"Gene Name\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene_expression_per_group(adata, \"Metadata_target\", \"DMSO\", 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for moa in adata.obs[\"Metadata_Mode_of_Action\"].unique():\n",
    "#     if moa == \"DMSO\":\n",
    "#         continue\n",
    "#     adata_small = adata[adata.obs[\"Metadata_Mode_of_Action\"].isin([moa, \"DMSO\"])].copy()\n",
    "#     if adata_small.obs[\"Metadata_target\"].nunique() < 2:\n",
    "#         continue\n",
    "#     print(moa)\n",
    "#     adata_small.obs.loc[adata_small.obs[\"Metadata_Mode_of_Action\"] == \"DMSO\", \"Metadata_target\"] = \"DMSO\"\n",
    "#     print(adata_small.obs[\"Metadata_target\"].value_counts())\n",
    "#     gene_expression_per_group(adata_small, \"Metadata_target\", \"DMSO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1440, 12205)\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(\"full_raw_matrix_and_meta.csv\", index_col=0)\n",
    "leiden_dict = pickle.load(open(\"leiden_dict_lyon.pkl\", \"rb\"))\n",
    "df_meta = pd.read_csv(\"transcriptomics_moa_target 1.csv\", index_col=0)\n",
    "df_meta.loc[df_meta[\"Pathway / Target\"] == \"DMSO\", \"MoA_tr\"] = \"DMSO\"\n",
    "df_meta.loc[df_meta[\"Pathway / Target\"] == \"DMSO\", \"MoA_tr\"] = \"DMSO\"\n",
    "# df_meta.columns = [ \"Metadata_\" + x for x in df_meta.columns]\n",
    "moa_dict = dict(zip(df_meta[\"BCS-code\"], df_meta[\"MoA_tr\"]))\n",
    "target_dict = dict(zip(df_meta[\"BCS-code\"], df_meta[\"MoA_tr\"]))\n",
    "moa_dict[\"DMSO\"] = \"DMSO\"\n",
    "target_dict[\"DMSO\"] = \"DMSO\"\n",
    "\n",
    "df[\"Metadata_MoA\"] = np.nan\n",
    "df[\"Metadata_target\"] = np.nan\n",
    "df[\"Metadata_MOA\"] = df[\"Metadata_treatments\"].map(moa_dict)\n",
    "df[\"Metadata_target\"] = df[\"Metadata_treatments\"].map(target_dict)\n",
    "df[\"Metadata_leiden\"] = df[\"Metadata_treatments\"].map(leiden_dict)\n",
    "\n",
    "gene_mapping = pickle.load(open(\"gene_mapping_botrytis.pkl\", \"rb\"))\n",
    "print(df.shape)\n",
    "df.rename(\n",
    "    columns={\n",
    "        \"Metadata_MOA\": \"Metadata_Mode_of_Action\",\n",
    "        \"Metadata_treatments\": \"Metadata_treatments\",\n",
    "    },\n",
    "    inplace=True,\n",
    ")\n",
    "df.loc[df[\"Metadata_treatments\"] == \"DMSO\", \"Metadata_concentration\"] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Leiden cluster 3', 'Leiden cluster 6', 'Leiden cluster 5',\n",
       "       'Leiden cluster 10', 'Leiden cluster 20', 'Leiden cluster 15',\n",
       "       'Leiden cluster 0', 'Leiden cluster 8', 'Leiden cluster 13',\n",
       "       'Leiden cluster 17', 'Leiden cluster 1', 'Leiden cluster 7',\n",
       "       'Leiden cluster 12', 'Leiden cluster 14', 'Leiden cluster 16',\n",
       "       'Leiden cluster 2', 'Leiden cluster 19', 'Leiden cluster 18',\n",
       "       'Leiden cluster 11', 'Leiden cluster 4', 'Leiden cluster 9', nan],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"Metadata_leiden\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "s = df.columns.to_series().groupby(df.columns)\n",
    "\n",
    "df.columns = np.where(\n",
    "    s.transform(\"size\") > 1, df.columns + s.cumcount().add(1).astype(str), df.columns\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: count, dtype: int64)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_name_counts = df.columns.value_counts()\n",
    "gene_name_counts[gene_name_counts > 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.default_inference import DefaultInference\n",
    "from pydeseq2.ds import DeseqStats\n",
    "\n",
    "\n",
    "def deseq(df: pd.DataFrame, group: str, reference: str) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Perform differential expression analysis using DESeq2.\n",
    "\n",
    "    This function takes a DataFrame containing gene expression counts and metadata,\n",
    "    and performs differential expression analysis using the DESeq2 method. It prepares\n",
    "    the data, fits the DESeq2 model, and returns a summary of the results sorted by\n",
    "    adjusted p-values.\n",
    "\n",
    "    Parameters:\n",
    "    ----------\n",
    "    df : pandas.DataFrame\n",
    "        A DataFrame where rows represent genes and columns represent samples. The DataFrame\n",
    "        should include metadata columns prefixed with \"Metadata_\" and a column for the\n",
    "        grouping variable.\n",
    "\n",
    "    group : str\n",
    "        The name of the column in the DataFrame that specifies the grouping variable for\n",
    "        the differential expression analysis. This variable should contain the treatment\n",
    "        groups to be compared.\n",
    "\n",
    "    reference : str\n",
    "        The reference level for the contrast in the differential expression analysis. This\n",
    "        should be one of the levels present in the `group` column.\n",
    "\n",
    "    Returns:\n",
    "    -------\n",
    "    pandas.DataFrame\n",
    "        A DataFrame containing the results of the differential expression analysis, sorted\n",
    "        by adjusted p-values.\n",
    "    \"\"\"\n",
    "\n",
    "    # if os.path.isfile(f\"differential_expression_tables/{treatment}_differential_expression.csv\"):\n",
    "    #     print(\"File exists, skipping\")\n",
    "    #     return\n",
    "    # Identify gene columns by excluding metadata columns\n",
    "    gene_columns = df.columns[~df.columns.str.startswith(\"Metadata_\")]\n",
    "\n",
    "    # Drop rows with NA values in specified columns\n",
    "    df.dropna(subset=[\"Metadata_plate\", group], inplace=True)\n",
    "\n",
    "    # Prepare counts DataFrame with genes as rows and samples as columns, filling NA with 0\n",
    "    counts = df[gene_columns].fillna(0).astype(int)\n",
    "\n",
    "    # Retain only genes with non-zero variance\n",
    "    counts = counts.loc[:, counts.var() != 0]\n",
    "\n",
    "    # Prepare metadata DataFrame\n",
    "    metadata = df[[\"Metadata_plate\", group]]\n",
    "    print(counts.shape)  # Print shape of counts DataFrame\n",
    "    print(metadata.shape)  # Print shape of metadata DataFrame\n",
    "\n",
    "    # Initialize inference with specified number of CPUs\n",
    "    inference = DefaultInference(n_cpus=14)\n",
    "\n",
    "    # Create DESeqDataSet object\n",
    "    dds = DeseqDataSet(\n",
    "        counts=counts.reset_index(drop=True),\n",
    "        metadata=metadata.reset_index(drop=True),\n",
    "        design=f\"~ Metadata_plate + {group}\",\n",
    "        refit_cooks=True,\n",
    "        inference=inference,\n",
    "    )\n",
    "\n",
    "    # Fit size factors and dispersions\n",
    "    dds.fit_size_factors()\n",
    "    dds.fit_genewise_dispersions()\n",
    "    dds.fit_dispersion_trend()\n",
    "    dds.fit_dispersion_prior()\n",
    "    dds.fit_MAP_dispersions()\n",
    "    dds.fit_LFC()\n",
    "    dds.calculate_cooks()\n",
    "\n",
    "    # Refitting outlier counts if specified\n",
    "    if dds.refit_cooks:\n",
    "        dds.refit()\n",
    "\n",
    "    # Set up and run differential expression statistics\n",
    "    ds = DeseqStats(\n",
    "        dds,\n",
    "        contrast=[group, treatment, reference],\n",
    "        alpha=0.05,\n",
    "        cooks_filter=True,\n",
    "        independent_filter=True,\n",
    "    )\n",
    "    ds.run_wald_test()\n",
    "\n",
    "    # Perform independent filtering or p-value adjustment\n",
    "    if ds.independent_filter:\n",
    "        ds._independent_filtering()\n",
    "    else:\n",
    "        ds._p_value_adjustment()\n",
    "\n",
    "    ds.summary()  # Print summary of results\n",
    "\n",
    "    # Sort results by adjusted p-values\n",
    "    summary_df = ds.results_df.copy()\n",
    "    summary_df.sort_values(\"padj\", inplace=True)\n",
    "\n",
    "    return summary_df  # Return sorted results DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from adjustText import adjust_text\n",
    "import os\n",
    "\n",
    "\n",
    "def make_volcano(data_frame, treatment):\n",
    "    \"\"\"\n",
    "    Make an interactive volcano plot as an html. Attach meta data to individual genes.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    treatment : string\n",
    "        Name of the treatment\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    fig : object\n",
    "        The figure object of the volcano plot\n",
    "\n",
    "    pd.Series : object\n",
    "        Pandas series containing the log2FoldChange. Will be used later for\n",
    "        colouring in the plot_gene_atlas function\n",
    "    \"\"\"\n",
    "    # if os.path.isfile(f\"differential_expression_tables/{treatment}_differential_expression.csv\"):\n",
    "    #     print(\"File exists, skipping\")\n",
    "    #     return\n",
    "\n",
    "\n",
    "    # Convert p-values into -Log10p\n",
    "    data_frame[\"-Log10P\"] = np.log10(data_frame[\"padj\"]) * -1\n",
    "    data_frame.loc[data_frame[\"-Log10P\"] < 0, \"-Log10P\"] = 0\n",
    "\n",
    "    # Filter for significant Up-regulated and Down-regulated genes\n",
    "    data_frame[\"Type hit\"] = \"Under Threshold\"\n",
    "    data_frame.loc[\n",
    "        (data_frame[\"log2FoldChange\"] > 1) & (data_frame[\"padj\"] < 0.05), \"Type hit\"\n",
    "    ] = \"Up-regulated over threshold of P < 0.05 \\n and log(2) Fold-Change > 1\"\n",
    "    data_frame.loc[\n",
    "        (data_frame[\"log2FoldChange\"] < -1) & (data_frame[\"padj\"] < 0.05), \"Type hit\"\n",
    "    ] = \"Down-regulated over threshold of P < 0.05 \\n and log(2) Fold-Change > 1\"\n",
    "\n",
    "    # Filter out non-significant genes\n",
    "    data_frame.sort_values(\"-Log10P\", ascending=False, inplace=True)\n",
    "    print(data_frame.head())\n",
    "    data_frame.to_csv(f\"{treatment}_differential_expression.csv\")\n",
    "\n",
    "    data_frame.rename(\n",
    "        columns={\n",
    "            \"log2FoldChange\": \"Log(2) Fold-Change\",\n",
    "            \"-Log10P\": \"-Log(10) of adjusted P-value\",\n",
    "        },\n",
    "        inplace=True,\n",
    "    )\n",
    "\n",
    "    all_significant = data_frame[\n",
    "        data_frame[\"Type hit\"] != \"Under Threshold\"\n",
    "    ].sort_values(\"-Log(10) of adjusted P-value\", ascending=False)\n",
    "    filtered_data = all_significant.iloc[:15]\n",
    "\n",
    "    fig = plt.figure(figsize=(18, 12))\n",
    "\n",
    "    colour_dict = {\n",
    "        \"Under Threshold\": \"lightgrey\",\n",
    "        \"Up-regulated over threshold of P < 0.05 \\n and log(2) Fold-Change > 1\": \"cornflowerblue\",\n",
    "        \"Down-regulated over threshold of P < 0.05 \\n and log(2) Fold-Change > 1\": \"crimson\",\n",
    "    }\n",
    "\n",
    "    # Make figure\n",
    "    g = sns.scatterplot(\n",
    "        data=data_frame,\n",
    "        x=\"Log(2) Fold-Change\",\n",
    "        y=\"-Log(10) of adjusted P-value\",\n",
    "        hue=\"Type hit\",\n",
    "        alpha=0.6,\n",
    "        palette=colour_dict,\n",
    "        legend=True,\n",
    "        edgecolor=\"none\",\n",
    "    )\n",
    "\n",
    "    # Add lines to show thresholds\n",
    "    ax = plt.gca()\n",
    "    ax.axhline(y=np.log10(0.05) * -1, lw=2, ls=\"--\", c=\"grey\")\n",
    "    ax.axvline(x=1, lw=2, ls=\"--\", c=\"grey\")\n",
    "    ax.axvline(x=-1, lw=2, ls=\"--\", c=\"grey\")\n",
    "    ax.spines[[\"right\", \"top\"]].set_visible(False)\n",
    "    ax.grid(True, alpha=0.2)\n",
    "    ax.tick_params(axis=\"both\", which=\"major\", labelsize=12)\n",
    "\n",
    "    texts = []\n",
    "    for gene, row in filtered_data.iterrows():\n",
    "        # Check the log2 fold change to determine text position\n",
    "        if row[\"Log(2) Fold-Change\"] < 1:\n",
    "            x_offset = 0.05  # Offset to the right\n",
    "        else:\n",
    "            x_offset = -0.05  # Offset to the left for better visibility\n",
    "\n",
    "        texts.append(\n",
    "            ax.text(\n",
    "                row[\"Log(2) Fold-Change\"] + x_offset,\n",
    "                row[\"-Log(10) of adjusted P-value\"] * 1.01,\n",
    "                f\"{gene_mapping[gene]}\",\n",
    "                fontsize=14,\n",
    "                color=\"black\",\n",
    "            )\n",
    "        )\n",
    "\n",
    "    # Use adjustText to avoid overlapping labels\n",
    "    adjust_text(\n",
    "        texts,\n",
    "        expand=(1.1, 1.1),\n",
    "        # only_move='y',  # Move only in the y direction\n",
    "        arrowprops=dict(color=\"silver\", lw=1),\n",
    "    )\n",
    "\n",
    "    sns.move_legend(ax, \"upper left\", bbox_to_anchor=(1, 1))\n",
    "    plt.title(f\"Volcano plot for {treatment} vs. DMSO\", fontsize=18)\n",
    "    plt.savefig(f\"volcano_{treatment}.png\")\n",
    "    plt.close()\n",
    "    # plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[~df[\"Metadata_leiden\"].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "group = \"Metadata_target\"\n",
    "group = \"Metadata_Mode_of_Action\"\n",
    "#group = \"Metadata_treatments\"\n",
    "group = \"Metadata_leiden\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "dmso_leidens = df.loc[df[\"Metadata_target\"] == \"DMSO\", \"Metadata_leiden\"].unique()\n",
    "dmso_leiden_dict = dict([(x, \"DMSO\") for x in dmso_leidens])\n",
    "df[\"Metadata_leiden\"] = df[\"Metadata_leiden\"].replace(dmso_leiden_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(200, 11746)\n",
      "(200, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.07 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 32.38 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.33 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 29.79 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.45 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.18 seconds.\n",
      "\n",
      "Replacing 7 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.06 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.04 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.21 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_target lipid other  vs DMSO\n",
      "                  baseMean  log2FoldChange     lfcSE      stat        pvalue  \\\n",
      "Bcin01g00010     24.830269       -2.251783  0.307282 -7.328071  2.334896e-13   \n",
      "Bcin01g00020     15.499798       -1.929769  0.326937 -5.902577  3.578668e-09   \n",
      "Bcin01g00030     81.027390       -2.625814  0.347506 -7.556165  4.151259e-14   \n",
      "Bcin01g00040     27.295532       -2.583620  0.336919 -7.668359  1.742103e-14   \n",
      "Bcin01g00050     14.464725       -1.373388  0.233373 -5.884955  3.981615e-09   \n",
      "...                    ...             ...       ...       ...           ...   \n",
      "ENSRNA049513861   0.044430        0.331456  4.959641  0.066831  9.467165e-01   \n",
      "ENSRNA049513883   0.347546       -0.283505  0.516350 -0.549056  5.829667e-01   \n",
      "ENSRNA049514038   0.215926       -0.612377  0.639439 -0.957678  3.382250e-01   \n",
      "ENSRNA049514141   0.104523        0.203421  1.717707  0.118426  9.057304e-01   \n",
      "ENSRNA049514255   0.220159       -0.450296  0.733859 -0.613601  5.394792e-01   \n",
      "\n",
      "                         padj  \n",
      "Bcin01g00010     3.297768e-12  \n",
      "Bcin01g00020     2.650763e-08  \n",
      "Bcin01g00030     6.437486e-13  \n",
      "Bcin01g00040     2.838478e-13  \n",
      "Bcin01g00050     2.921997e-08  \n",
      "...                       ...  \n",
      "ENSRNA049513861           NaN  \n",
      "ENSRNA049513883  6.778698e-01  \n",
      "ENSRNA049514038           NaN  \n",
      "ENSRNA049514141           NaN  \n",
      "ENSRNA049514255           NaN  \n",
      "\n",
      "[11746 rows x 6 columns]\n",
      "                 baseMean  log2FoldChange     lfcSE       stat        pvalue  \\\n",
      "Bcin06g01180  1513.189209        5.854483  0.300781  19.464279  2.205765e-84   \n",
      "Bcin07g00780  1781.956909        2.765313  0.153282  18.040709  9.333850e-73   \n",
      "Bcin06g05330   220.940491        5.525952  0.311650  17.731294  2.404353e-70   \n",
      "Bcin02g04960    93.995857        2.729058  0.157419  17.336292  2.503791e-67   \n",
      "Bcin14g04420   222.077225        3.591832  0.208313  17.242449  1.275180e-66   \n",
      "\n",
      "                      padj    -Log10P  \\\n",
      "Bcin06g01180  2.339655e-80  79.630848   \n",
      "Bcin07g00780  4.950207e-69  68.305377   \n",
      "Bcin06g05330  8.500992e-67  66.070530   \n",
      "Bcin02g04960  6.639428e-64  63.177869   \n",
      "Bcin14g04420  2.705167e-63  62.567806   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin06g01180  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin07g00780  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin06g05330  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin02g04960  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin14g04420  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "(320, 11756)\n",
      "(320, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.09 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 46.74 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.37 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 43.34 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.89 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.28 seconds.\n",
      "\n",
      "Replacing 26 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.14 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.09 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.15 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_target lipid SBI  vs DMSO\n",
      "                   baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "Bcin01g00010      67.949715        0.879477  0.296370  2.967495  0.003002   \n",
      "Bcin01g00020      43.114182        0.828876  0.286293  2.895199  0.003789   \n",
      "Bcin01g00030     211.640091        0.641467  0.280848  2.284040  0.022369   \n",
      "Bcin01g00040      70.952797        0.698617  0.284455  2.455980  0.014050   \n",
      "Bcin01g00050      28.541906        0.575963  0.234457  2.456582  0.014027   \n",
      "...                     ...             ...       ...       ...       ...   \n",
      "ENSRNA049513861    0.030497        0.267874  4.465506  0.059987  0.952166   \n",
      "ENSRNA049513883    0.430430        0.052329  0.326812  0.160120  0.872787   \n",
      "ENSRNA049514038    0.369934        0.287693  0.400914  0.717593  0.473008   \n",
      "ENSRNA049514141    0.205839        0.344642  0.766075  0.449880  0.652797   \n",
      "ENSRNA049514255    0.279365       -0.123719  0.485880 -0.254628  0.799011   \n",
      "\n",
      "                     padj  \n",
      "Bcin01g00010     0.013077  \n",
      "Bcin01g00020     0.015680  \n",
      "Bcin01g00030     0.063359  \n",
      "Bcin01g00040     0.044614  \n",
      "Bcin01g00050     0.044554  \n",
      "...                   ...  \n",
      "ENSRNA049513861       NaN  \n",
      "ENSRNA049513883       NaN  \n",
      "ENSRNA049514038       NaN  \n",
      "ENSRNA049514141       NaN  \n",
      "ENSRNA049514255       NaN  \n",
      "\n",
      "[11756 rows x 6 columns]\n",
      "                baseMean  log2FoldChange     lfcSE       stat        pvalue  \\\n",
      "Bcin14g02510  515.653503        2.687942  0.193436  13.895735  6.723003e-44   \n",
      "Bcin06g07150   50.704441        4.813430  0.408967  11.769723  5.590765e-32   \n",
      "Bcin14g05330   32.247570        4.037554  0.350528  11.518492  1.064578e-30   \n",
      "Bcin01g01260  965.634460        5.541435  0.483248  11.467060  1.931099e-30   \n",
      "Bcin04g04450  444.364105        2.438945  0.214260  11.383129  5.074554e-30   \n",
      "\n",
      "                      padj    -Log10P  \\\n",
      "Bcin14g02510  6.524002e-40  39.185486   \n",
      "Bcin06g07150  2.712639e-28  27.566608   \n",
      "Bcin14g05330  3.443555e-27  26.462993   \n",
      "Bcin01g01260  4.684847e-27  26.329305   \n",
      "Bcin04g04450  9.848694e-27  26.006621   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin14g02510  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin06g07150  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin14g05330  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin01g01260  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin04g04450  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "(130, 11739)\n",
      "(130, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.05 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 23.24 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.28 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 22.16 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.24 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.10 seconds.\n",
      "\n",
      "Replacing 7 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.04 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.25 seconds.\n",
      "\n",
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/pandas/core/arraylike.py:396: RuntimeWarning: divide by zero encountered in log10\n",
      "  result = getattr(ufunc, method)(*inputs, **kwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_target prot metabolism  vs DMSO\n",
      "                  baseMean  log2FoldChange     lfcSE       stat        pvalue  \\\n",
      "Bcin01g00010     16.631866       -2.962161  0.220871 -13.411243  5.196066e-41   \n",
      "Bcin01g00020     22.143028        0.448576  0.253305   1.770891  7.657876e-02   \n",
      "Bcin01g00030     61.338985       -3.265744  0.238265 -13.706359  9.300982e-43   \n",
      "Bcin01g00040     21.362955       -3.349372  0.267873 -12.503600  7.134529e-36   \n",
      "Bcin01g00050     11.071600       -1.419013  0.214069  -6.628770  3.384945e-11   \n",
      "...                    ...             ...       ...        ...           ...   \n",
      "ENSRNA049513861   0.049370        1.888931  4.295027   0.439795  6.600857e-01   \n",
      "ENSRNA049513883   0.955641        2.306129  0.541292   4.260413  2.040496e-05   \n",
      "ENSRNA049514038   0.228123        0.567258  1.021743   0.555187  5.787667e-01   \n",
      "ENSRNA049514141   0.167177        1.632212  2.538605   0.642956  5.202526e-01   \n",
      "ENSRNA049514255   1.474402        3.504414  0.604474   5.797457  6.732790e-09   \n",
      "\n",
      "                         padj  \n",
      "Bcin01g00010     5.032725e-40  \n",
      "Bcin01g00020     9.818240e-02  \n",
      "Bcin01g00030     9.705264e-42  \n",
      "Bcin01g00040     5.756167e-35  \n",
      "Bcin01g00050     8.613889e-11  \n",
      "...                       ...  \n",
      "ENSRNA049513861  6.905151e-01  \n",
      "ENSRNA049513883  3.602554e-05  \n",
      "ENSRNA049514038  6.288544e-01  \n",
      "ENSRNA049514141  5.710374e-01  \n",
      "ENSRNA049514255  1.497182e-08  \n",
      "\n",
      "[11739 rows x 6 columns]\n",
      "                 baseMean  log2FoldChange     lfcSE       stat         pvalue  \\\n",
      "Bcin07g04780   939.133057        5.244169  0.124265  42.201356   0.000000e+00   \n",
      "Bcin12g01070   643.627075        6.271021  0.164711  38.072780   0.000000e+00   \n",
      "Bcin12g02800  3900.528809        7.046120  0.190771  36.934979  1.269121e-298   \n",
      "Bcin04g01520  1458.245972        6.273075  0.170118  36.874811  1.170999e-297   \n",
      "Bcin12g05660   646.412964        5.262586  0.145131  36.261034  6.658673e-288   \n",
      "\n",
      "                       padj     -Log10P  \\\n",
      "Bcin07g04780   0.000000e+00         inf   \n",
      "Bcin12g01070   0.000000e+00         inf   \n",
      "Bcin12g02800  4.966071e-295  294.303987   \n",
      "Bcin04g01520  3.436590e-294  293.463872   \n",
      "Bcin12g05660  1.563323e-284  283.805951   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin07g04780  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin12g01070  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin12g02800  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin04g01520  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin12g05660  Up-regulated over threshold of P < 0.05 \\n and...  \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/matplotlib/transforms.py:360: RuntimeWarning: invalid value encountered in scalar subtract\n",
      "  return points[1, 1] - points[0, 1]\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(255, 11754)\n",
      "(255, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.09 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 38.25 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.34 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 37.13 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.49 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.21 seconds.\n",
      "\n",
      "Replacing 43 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.19 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.15 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.23 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_target respiration  vs DMSO\n",
      "                   baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "Bcin01g00010      69.545990        0.334474  0.278560  1.200726  0.229858   \n",
      "Bcin01g00020      39.887661        0.457207  0.257903  1.772785  0.076264   \n",
      "Bcin01g00030     220.220932        0.237123  0.269070  0.881269  0.378172   \n",
      "Bcin01g00040      68.967293        0.253373  0.274167  0.924155  0.355405   \n",
      "Bcin01g00050      30.632471        0.486086  0.206328  2.355891  0.018478   \n",
      "...                     ...             ...       ...       ...       ...   \n",
      "ENSRNA049513861    0.032574        0.234466  4.074557  0.057544  0.954112   \n",
      "ENSRNA049513883    0.564496        0.337406  0.320189  1.053771  0.291988   \n",
      "ENSRNA049514038    0.194776       -0.574689  0.550264 -1.044388  0.296306   \n",
      "ENSRNA049514141    0.105885        0.195382  0.978977  0.199577  0.841811   \n",
      "ENSRNA049514255    0.226951        0.012547  0.509440  0.024629  0.980351   \n",
      "\n",
      "                     padj  \n",
      "Bcin01g00010     0.314193  \n",
      "Bcin01g00020     0.123192  \n",
      "Bcin01g00030     0.471394  \n",
      "Bcin01g00040     0.447748  \n",
      "Bcin01g00050     0.035524  \n",
      "...                   ...  \n",
      "ENSRNA049513861       NaN  \n",
      "ENSRNA049513883  0.380486  \n",
      "ENSRNA049514038       NaN  \n",
      "ENSRNA049514141       NaN  \n",
      "ENSRNA049514255       NaN  \n",
      "\n",
      "[11754 rows x 6 columns]\n",
      "                 baseMean  log2FoldChange     lfcSE       stat        pvalue  \\\n",
      "Bcin16g00180    57.698769       -0.477250  0.024983 -19.102657  2.399606e-81   \n",
      "Bcin13g03160   440.804504       -2.087415  0.110038 -18.969959  3.021432e-80   \n",
      "Bcin08g01540   454.671997       -1.135108  0.061225 -18.540069  9.809647e-77   \n",
      "Bcin12g06180  1437.132080        2.827921  0.167093  16.924217  2.982851e-64   \n",
      "Bcin09g02600   530.888428        1.128074  0.070252  16.057547  5.061803e-58   \n",
      "\n",
      "                      padj    -Log10P  \\\n",
      "Bcin16g00180  2.546942e-77  76.593981   \n",
      "Bcin13g03160  1.603474e-76  75.794938   \n",
      "Bcin08g01540  3.470653e-73  72.459589   \n",
      "Bcin12g06180  7.914994e-61  60.101549   \n",
      "Bcin09g02600  1.074520e-54  53.968786   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin16g00180                                    Under Threshold  \n",
      "Bcin13g03160  Down-regulated over threshold of P < 0.05 \\n a...  \n",
      "Bcin08g01540  Down-regulated over threshold of P < 0.05 \\n a...  \n",
      "Bcin12g06180  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin09g02600  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "(90, 11707)\n",
      "(90, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 20.87 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.39 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 19.07 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.32 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.06 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.18 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_target IMPDH vs DMSO\n",
      "                   baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "Bcin01g00010      40.512295       -0.100411  0.270379 -0.371373  0.710360   \n",
      "Bcin01g00020      30.553429        0.194323  0.269621  0.720726  0.471078   \n",
      "Bcin01g00030     153.941910       -0.333775  0.262643 -1.270831  0.203789   \n",
      "Bcin01g00040      57.920860       -0.128718  0.265800 -0.484264  0.628198   \n",
      "Bcin01g00050      22.829159       -0.047238  0.212140 -0.222671  0.823791   \n",
      "...                     ...             ...       ...       ...       ...   \n",
      "ENSRNA049513861    0.057758        1.125603  3.555680  0.316565  0.751574   \n",
      "ENSRNA049513883    0.404849        0.608123  0.595871  1.020563  0.307462   \n",
      "ENSRNA049514038    0.204051        0.708808  1.004354  0.705735  0.480353   \n",
      "ENSRNA049514141    0.129058        1.267291  1.875954  0.675545  0.499330   \n",
      "ENSRNA049514255    0.176431        0.427736  1.097137  0.389865  0.696636   \n",
      "\n",
      "                     padj  \n",
      "Bcin01g00010     0.796383  \n",
      "Bcin01g00020     0.600753  \n",
      "Bcin01g00030     0.329131  \n",
      "Bcin01g00040     0.732290  \n",
      "Bcin01g00050     0.877173  \n",
      "...                   ...  \n",
      "ENSRNA049513861       NaN  \n",
      "ENSRNA049513883  0.445164  \n",
      "ENSRNA049514038  0.609058  \n",
      "ENSRNA049514141  0.624077  \n",
      "ENSRNA049514255  0.785791  \n",
      "\n",
      "[11707 rows x 6 columns]\n",
      "                baseMean  log2FoldChange     lfcSE       stat        pvalue  \\\n",
      "Bcin12g05690  131.678665        2.733666  0.159800  17.106820  1.320131e-65   \n",
      "Bcin15g00170  694.413513        8.164059  0.496741  16.435236  1.070066e-60   \n",
      "Bcin07g04950  381.810730        3.184665  0.201314  15.819358  2.288140e-56   \n",
      "Bcin14g02510  236.029312        2.725261  0.175410  15.536547  1.962776e-54   \n",
      "Bcin12g00520  116.545441        7.301244  0.474605  15.383833  2.101319e-53   \n",
      "\n",
      "                      padj    -Log10P  \\\n",
      "Bcin12g05690  1.455577e-61  60.836965   \n",
      "Bcin15g00170  5.899272e-57  56.229202   \n",
      "Bcin07g04950  8.409678e-53  52.075221   \n",
      "Bcin14g02510  5.410392e-51  50.266771   \n",
      "Bcin12g00520  4.633829e-50  49.334060   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin12g05690  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin15g00170  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin07g04950  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin14g02510  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin12g00520  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "(115, 11727)\n",
      "(115, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.05 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 23.65 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.32 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 23.66 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.17 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.09 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.22 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_target cell wal  vs DMSO\n",
      "                   baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "Bcin01g00010      43.905602       -0.341886  0.232916 -1.467853  0.142144   \n",
      "Bcin01g00020      33.583134       -0.178443  0.223243 -0.799321  0.424104   \n",
      "Bcin01g00030     188.741241       -0.273665  0.245187 -1.116146  0.264360   \n",
      "Bcin01g00040      71.145363       -0.216734  0.245747 -0.881940  0.377809   \n",
      "Bcin01g00050      28.406347       -0.233789  0.196688 -1.188628  0.234586   \n",
      "...                     ...             ...       ...       ...       ...   \n",
      "ENSRNA049513861    0.056467       -0.238441  3.924049 -0.060764  0.951547   \n",
      "ENSRNA049513883    0.550826       -0.156983  0.468950 -0.334754  0.737810   \n",
      "ENSRNA049514038    0.252265        0.313915  0.871099  0.360367  0.718573   \n",
      "ENSRNA049514141    0.147235       -0.027341  1.959475 -0.013953  0.988867   \n",
      "ENSRNA049514255    0.197416       -0.586398  0.947954 -0.618594  0.536184   \n",
      "\n",
      "                     padj  \n",
      "Bcin01g00010     0.327911  \n",
      "Bcin01g00020     0.632359  \n",
      "Bcin01g00030     0.482884  \n",
      "Bcin01g00040     0.593411  \n",
      "Bcin01g00050     0.447579  \n",
      "...                   ...  \n",
      "ENSRNA049513861       NaN  \n",
      "ENSRNA049513883       NaN  \n",
      "ENSRNA049514038       NaN  \n",
      "ENSRNA049514141       NaN  \n",
      "ENSRNA049514255       NaN  \n",
      "\n",
      "[11727 rows x 6 columns]\n",
      "                baseMean  log2FoldChange     lfcSE       stat        pvalue  \\\n",
      "Bcin06g01180  195.600800        3.549294  0.302210  11.744478  7.538695e-32   \n",
      "Bcin08g06430  222.545517       -0.217796  0.020121 -10.824163  2.644855e-27   \n",
      "Bcin14g03140   67.004700        1.779288  0.169950  10.469497  1.192749e-25   \n",
      "Bcin01g05880  113.758820        5.996107  0.608231   9.858274  6.312514e-23   \n",
      "Bcin03g03910   54.880070        0.425189  0.043456   9.784349  1.314388e-22   \n",
      "\n",
      "                      padj    -Log10P  \\\n",
      "Bcin06g01180  7.469339e-28  27.126718   \n",
      "Bcin08g06430  1.310261e-23  22.882642   \n",
      "Bcin14g03140  3.939254e-22  21.404586   \n",
      "Bcin01g05880  1.563610e-19  18.805872   \n",
      "Bcin03g03910  2.604592e-19  18.584260   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin06g01180  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin08g06430                                    Under Threshold  \n",
      "Bcin14g03140  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin01g05880  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin03g03910                                    Under Threshold  \n",
      "(100, 11716)\n",
      "(100, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.04 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 22.01 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.38 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 20.18 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.23 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.07 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.22 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_target Cell wall  vs DMSO\n",
      "                   baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "Bcin01g00010      48.671753        0.188198  0.202956  0.927285  0.353778   \n",
      "Bcin01g00020      41.225513        0.553291  0.207895  2.661399  0.007782   \n",
      "Bcin01g00030     220.398911        0.450815  0.194781  2.314469  0.020642   \n",
      "Bcin01g00040      78.731827        0.384636  0.181867  2.114932  0.034436   \n",
      "Bcin01g00050      30.633778        0.212699  0.167520  1.269694  0.204194   \n",
      "...                     ...             ...       ...       ...       ...   \n",
      "ENSRNA049513861    0.063577       -0.145355  3.918647 -0.037093  0.970411   \n",
      "ENSRNA049513883    0.534639        0.150656  0.485173  0.310521  0.756165   \n",
      "ENSRNA049514038    0.338306        0.697529  0.827624  0.842809  0.399335   \n",
      "ENSRNA049514141    0.163462        0.137426  1.823251  0.075374  0.939917   \n",
      "ENSRNA049514255    0.203255       -0.499966  1.016346 -0.491925  0.622773   \n",
      "\n",
      "                     padj  \n",
      "Bcin01g00010     0.638422  \n",
      "Bcin01g00020     0.066063  \n",
      "Bcin01g00030     0.125567  \n",
      "Bcin01g00040     0.174381  \n",
      "Bcin01g00050     0.489043  \n",
      "...                   ...  \n",
      "ENSRNA049513861       NaN  \n",
      "ENSRNA049513883       NaN  \n",
      "ENSRNA049514038       NaN  \n",
      "ENSRNA049514141       NaN  \n",
      "ENSRNA049514255       NaN  \n",
      "\n",
      "[11716 rows x 6 columns]\n",
      "                 baseMean  log2FoldChange     lfcSE       stat        pvalue  \\\n",
      "Bcin14g02510   227.929428        2.535962  0.131749  19.248429  1.455086e-82   \n",
      "Bcin05g03360    25.691628        2.271075  0.189308  11.996720  3.696581e-33   \n",
      "Bcin08g07120    63.728218        2.000950  0.189043  10.584626  3.511801e-26   \n",
      "Bcin08g06710  2138.678223        0.861883  0.084509  10.198763  2.008134e-24   \n",
      "Bcin14g05330    11.706299        3.551219  0.359838   9.868946  5.675779e-23   \n",
      "\n",
      "                      padj    -Log10P  \\\n",
      "Bcin14g02510  1.407213e-78  77.851640   \n",
      "Bcin05g03360  1.787482e-29  28.747758   \n",
      "Bcin08g07120  1.132088e-22  21.946120   \n",
      "Bcin08g06710  4.855166e-21  20.313796   \n",
      "Bcin14g05330  1.097809e-19  18.959473   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin14g02510  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin05g03360  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin08g07120  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin08g06710                                    Under Threshold  \n",
      "Bcin14g05330  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "(165, 11737)\n",
      "(165, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.05 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 30.26 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.38 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 26.64 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.83 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.14 seconds.\n",
      "\n",
      "Replacing 6 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.04 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.21 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_target cytoskeleton  vs DMSO\n",
      "                  baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "Bcin01g00010     24.576702       -1.296579  0.343983 -3.769311  0.000164   \n",
      "Bcin01g00020     18.323969       -1.104318  0.325900 -3.388519  0.000703   \n",
      "Bcin01g00030     95.694176       -1.422698  0.351712 -4.045069  0.000052   \n",
      "Bcin01g00040     36.470383       -1.304039  0.351841 -3.706333  0.000210   \n",
      "Bcin01g00050     16.811169       -0.724881  0.197205 -3.675767  0.000237   \n",
      "...                    ...             ...       ...       ...       ...   \n",
      "ENSRNA049513861   0.038263        0.421258  4.261995  0.098841  0.921265   \n",
      "ENSRNA049513883   0.396675        0.232981  0.475203  0.490278  0.623937   \n",
      "ENSRNA049514038   0.342001        1.482361  0.680178  2.179373  0.029304   \n",
      "ENSRNA049514141   0.232137        1.208773  0.772575  1.564603  0.117676   \n",
      "ENSRNA049514255   0.107718       -0.011682  1.072294 -0.010895  0.991308   \n",
      "\n",
      "                     padj  \n",
      "Bcin01g00010     0.000968  \n",
      "Bcin01g00020     0.003400  \n",
      "Bcin01g00030     0.000354  \n",
      "Bcin01g00040     0.001200  \n",
      "Bcin01g00050     0.001324  \n",
      "...                   ...  \n",
      "ENSRNA049513861       NaN  \n",
      "ENSRNA049513883  0.750382  \n",
      "ENSRNA049514038  0.078671  \n",
      "ENSRNA049514141       NaN  \n",
      "ENSRNA049514255       NaN  \n",
      "\n",
      "[11737 rows x 6 columns]\n",
      "                baseMean  log2FoldChange     lfcSE       stat        pvalue  \\\n",
      "Bcin10g05060  149.553741        3.817689  0.250831  15.220163  2.598808e-52   \n",
      "Bcin05g01580  139.171707        2.917000  0.210265  13.872985  9.235209e-44   \n",
      "Bcin09g06840  234.812408        2.942576  0.225587  13.044076  6.868458e-39   \n",
      "Bcin03g02770  239.693985        3.623465  0.278775  12.997809  1.258991e-38   \n",
      "Bcin12g03840   63.632317        3.330790  0.260480  12.787146  1.934487e-37   \n",
      "\n",
      "                      padj    -Log10P  \\\n",
      "Bcin10g05060  2.754477e-48  47.559961   \n",
      "Bcin05g01580  4.894199e-40  39.310318   \n",
      "Bcin09g06840  2.426626e-35  34.614997   \n",
      "Bcin03g02770  3.336011e-35  34.476773   \n",
      "Bcin12g03840  4.100725e-34  33.387139   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin10g05060  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin05g01580  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin09g06840  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin03g02770  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin12g03840  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "(130, 11740)\n",
      "(130, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.04 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 26.50 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.34 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 23.97 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.38 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.10 seconds.\n",
      "\n",
      "Replacing 6 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.04 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.23 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_target DHODH vs DMSO\n",
      "                   baseMean  log2FoldChange     lfcSE      stat        pvalue  \\\n",
      "Bcin01g00010      33.369751       -1.465445  0.229400 -6.388154  1.679000e-10   \n",
      "Bcin01g00020      23.667122       -1.203882  0.233116 -5.164303  2.413368e-07   \n",
      "Bcin01g00030     128.198044       -1.529432  0.240748 -6.352838  2.113785e-10   \n",
      "Bcin01g00040      46.041630       -1.619297  0.240295 -6.738784  1.597172e-11   \n",
      "Bcin01g00050      21.880724       -0.794759  0.171965 -4.621638  3.807225e-06   \n",
      "...                     ...             ...       ...       ...           ...   \n",
      "ENSRNA049513861    0.049590        0.107845  4.003591  0.026937  9.785099e-01   \n",
      "ENSRNA049513883    0.393442       -0.307641  0.479704 -0.641314  5.213188e-01   \n",
      "ENSRNA049514038    0.257032        0.626508  0.847931  0.738866  4.599881e-01   \n",
      "ENSRNA049514141    0.161565        0.579184  0.957761  0.604727  5.453603e-01   \n",
      "ENSRNA049514255    0.164251       -0.242141  0.865729 -0.279696  7.797108e-01   \n",
      "\n",
      "                         padj  \n",
      "Bcin01g00010     3.640237e-09  \n",
      "Bcin01g00020     2.587111e-06  \n",
      "Bcin01g00030     4.464214e-09  \n",
      "Bcin01g00040     4.286890e-10  \n",
      "Bcin01g00050     3.090674e-05  \n",
      "...                       ...  \n",
      "ENSRNA049513861           NaN  \n",
      "ENSRNA049513883  6.630304e-01  \n",
      "ENSRNA049514038           NaN  \n",
      "ENSRNA049514141           NaN  \n",
      "ENSRNA049514255           NaN  \n",
      "\n",
      "[11740 rows x 6 columns]\n",
      "                baseMean  log2FoldChange     lfcSE       stat         pvalue  \\\n",
      "Bcin04g01890  295.169769       -0.465159  0.015741 -29.551077  6.361107e-192   \n",
      "Bcin03g04950  194.372025        0.390954  0.020001  19.546281   4.437096e-85   \n",
      "Bcin12g01410  499.241241        3.708314  0.209732  17.681235   5.850018e-70   \n",
      "Bcin03g00410   93.312393        4.203973  0.303531  13.850248   1.267735e-43   \n",
      "Bcin13g02590   77.038986        3.694706  0.269053  13.732267   6.506553e-43   \n",
      "\n",
      "                       padj     -Log10P  \\\n",
      "Bcin04g01890  6.744045e-188  187.171080   \n",
      "Bcin03g04950   2.352105e-81   80.628543   \n",
      "Bcin12g01410   2.067397e-66   65.684576   \n",
      "Bcin03g00410   3.360131e-40   39.473644   \n",
      "Bcin13g02590   1.379650e-39   38.860231   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin04g01890                                    Under Threshold  \n",
      "Bcin03g04950                                    Under Threshold  \n",
      "Bcin12g01410  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin03g00410  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin13g02590  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "(175, 11743)\n",
      "(175, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.06 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 29.54 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.40 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 27.62 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.37 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.14 seconds.\n",
      "\n",
      "Replacing 30 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.08 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.11 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.20 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_target GPI vs DMSO\n",
      "                   baseMean  log2FoldChange     lfcSE      stat        pvalue  \\\n",
      "Bcin01g00010      92.695740        1.694896  0.265265  6.389435  1.665003e-10   \n",
      "Bcin01g00020      72.386002        1.785965  0.246977  7.231315  4.783379e-13   \n",
      "Bcin01g00030     341.849121        1.473305  0.223381  6.595497  4.238346e-11   \n",
      "Bcin01g00040     124.292480        1.459724  0.247340  5.901689  3.597995e-09   \n",
      "Bcin01g00050      50.324451        1.442963  0.218185  6.613481  3.753856e-11   \n",
      "...                     ...             ...       ...       ...           ...   \n",
      "ENSRNA049513861    0.041646        0.246783  4.342075  0.056835  9.546764e-01   \n",
      "ENSRNA049513883    0.584550        0.437885  0.396159  1.105327  2.690177e-01   \n",
      "ENSRNA049514038    0.328520        1.212113  0.624237  1.941751  5.216722e-02   \n",
      "ENSRNA049514141    0.155371        0.546217  0.984166  0.555005  5.788911e-01   \n",
      "ENSRNA049514255    0.272433        0.754479  0.646893  1.166311  2.434888e-01   \n",
      "\n",
      "                         padj  \n",
      "Bcin01g00010     1.529490e-08  \n",
      "Bcin01g00020     1.078543e-10  \n",
      "Bcin01g00030     5.212857e-09  \n",
      "Bcin01g00040     1.994173e-07  \n",
      "Bcin01g00050     4.774616e-09  \n",
      "...                       ...  \n",
      "ENSRNA049513861           NaN  \n",
      "ENSRNA049513883           NaN  \n",
      "ENSRNA049514038           NaN  \n",
      "ENSRNA049514141           NaN  \n",
      "ENSRNA049514255           NaN  \n",
      "\n",
      "[11743 rows x 6 columns]\n",
      "                baseMean  log2FoldChange     lfcSE       stat        pvalue  \\\n",
      "Bcin14g02510  604.549988        3.725945  0.237417  15.693650  1.671606e-55   \n",
      "Bcin10g00050  103.741364        1.663980  0.153290  10.855146  1.885060e-27   \n",
      "Bcin08g07120  159.137619        3.331701  0.309042  10.780734  4.244857e-27   \n",
      "Bcin16g02140   22.193310        4.307048  0.402603  10.698003  1.039949e-26   \n",
      "Bcin02g05560  113.532280        4.159123  0.416371   9.988978  1.703295e-23   \n",
      "\n",
      "                      padj    -Log10P  \\\n",
      "Bcin14g02510  1.658401e-51  50.780311   \n",
      "Bcin10g00050  9.350841e-24  23.029149   \n",
      "Bcin08g07120  1.403774e-23  22.852703   \n",
      "Bcin16g02140  2.579334e-23  22.588492   \n",
      "Bcin02g05560  3.379678e-20  19.471125   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin14g02510  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin10g00050  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin08g07120  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin16g02140  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin02g05560  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "(100, 11727)\n",
      "(100, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.04 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 21.84 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.46 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 20.78 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.27 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.08 seconds.\n",
      "\n",
      "Replacing 1 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.26 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_target HDAC vs DMSO\n",
      "                   baseMean  log2FoldChange     lfcSE      stat        pvalue  \\\n",
      "Bcin01g00010      25.633125       -1.827829  0.251885 -7.256614  3.969003e-13   \n",
      "Bcin01g00020      19.758617       -1.468987  0.254313 -5.776303  7.635985e-09   \n",
      "Bcin01g00030     101.310417       -2.321982  0.258237 -8.991677  2.434868e-19   \n",
      "Bcin01g00040      36.823555       -2.210744  0.268229 -8.242006  1.693467e-16   \n",
      "Bcin01g00050      17.629440       -1.172277  0.213439 -5.492319  3.966897e-08   \n",
      "...                     ...             ...       ...       ...           ...   \n",
      "ENSRNA049513861    0.043816        0.896928  3.927959  0.228345  8.193784e-01   \n",
      "ENSRNA049513883    1.054442        0.459175  0.698112  0.657739  5.107061e-01   \n",
      "ENSRNA049514038    0.361117        2.278762  0.877750  2.596140  9.427767e-03   \n",
      "ENSRNA049514141    0.253538        0.909439  3.337276  0.272509  7.852304e-01   \n",
      "ENSRNA049514255    0.143633        0.899855  1.066096  0.844066  3.986328e-01   \n",
      "\n",
      "                         padj  \n",
      "Bcin01g00010     6.029081e-12  \n",
      "Bcin01g00020     6.188472e-08  \n",
      "Bcin01g00030     7.533957e-18  \n",
      "Bcin01g00040     3.789940e-15  \n",
      "Bcin01g00050     2.859238e-07  \n",
      "...                       ...  \n",
      "ENSRNA049513861  8.589301e-01  \n",
      "ENSRNA049513883  6.228214e-01  \n",
      "ENSRNA049514038  2.160630e-02  \n",
      "ENSRNA049514141  8.381497e-01  \n",
      "ENSRNA049514255  5.112946e-01  \n",
      "\n",
      "[11727 rows x 6 columns]\n",
      "                baseMean  log2FoldChange     lfcSE       stat         pvalue  \\\n",
      "Bcin09g03080  387.085693        7.004789  0.276592  25.325368  1.679033e-141   \n",
      "Bcin07g04840   66.804443        7.410334  0.300468  24.662671  2.691245e-134   \n",
      "Bcin12g02050   96.199608        2.820147  0.114520  24.625829  6.682162e-134   \n",
      "Bcin12g05690  127.079414        2.998517  0.122709  24.435972  7.094685e-132   \n",
      "Bcin04g01360   56.913029        6.109269  0.274302  22.272062  6.895485e-110   \n",
      "\n",
      "                       padj     -Log10P  \\\n",
      "Bcin09g03080  1.969002e-137  136.705754   \n",
      "Bcin07g04840  1.578011e-130  129.801890   \n",
      "Bcin12g02050  2.612057e-130  129.583017   \n",
      "Bcin12g05690  2.079984e-128  127.681940   \n",
      "Bcin04g01360  1.617267e-106  105.791218   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin09g03080  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin07g04840  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin12g02050  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin12g05690  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin04g01360  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "(110, 11714)\n",
      "(110, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 22.91 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.41 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 21.08 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.66 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.08 seconds.\n",
      "\n",
      "Replacing 17 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.05 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.24 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_target HOG vs DMSO\n",
      "                  baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "Bcin01g00010     15.540810       -0.128263  0.276844 -0.463304  0.643146   \n",
      "Bcin01g00020     11.707269       -0.079172  0.286508 -0.276334  0.782292   \n",
      "Bcin01g00030     59.017170       -0.577087  0.257079 -2.244780  0.024782   \n",
      "Bcin01g00040     22.916466       -0.526460  0.274393 -1.918635  0.055031   \n",
      "Bcin01g00050     10.734890        0.049900  0.244579  0.204025  0.838334   \n",
      "...                    ...             ...       ...       ...       ...   \n",
      "ENSRNA049513861   0.037536        3.967910  3.939424  1.007231  0.313824   \n",
      "ENSRNA049513883   0.196723        2.022989  0.657090  3.078709  0.002079   \n",
      "ENSRNA049514038   0.115538        3.560255  1.152829  3.088277  0.002013   \n",
      "ENSRNA049514141   0.323182        3.632094  3.541513  1.025577  0.305091   \n",
      "ENSRNA049514255   0.079593        3.005960  1.160980  2.589157  0.009621   \n",
      "\n",
      "                     padj  \n",
      "Bcin01g00010     0.698870  \n",
      "Bcin01g00020     0.821494  \n",
      "Bcin01g00030     0.045359  \n",
      "Bcin01g00040     0.091152  \n",
      "Bcin01g00050     0.867820  \n",
      "...                   ...  \n",
      "ENSRNA049513861  0.398626  \n",
      "ENSRNA049513883  0.005030  \n",
      "ENSRNA049514038  0.004896  \n",
      "ENSRNA049514141  0.392347  \n",
      "ENSRNA049514255  0.019546  \n",
      "\n",
      "[11714 rows x 6 columns]\n",
      "                baseMean  log2FoldChange     lfcSE       stat        pvalue  \\\n",
      "Bcin06g01180  568.394226        6.522920  0.308020  21.176953  1.557899e-99   \n",
      "Bcin15g00270   22.732100        5.850155  0.280041  20.890320  6.557875e-97   \n",
      "Bcin06g05330   37.652317        5.012690  0.242188  20.697485  3.649234e-95   \n",
      "Bcin02g08770  176.034943        6.274400  0.310076  20.235061  4.810298e-91   \n",
      "Bcin08g07120  127.996948        4.817730  0.248811  19.363022  1.583189e-83   \n",
      "\n",
      "                      padj    -Log10P  \\\n",
      "Bcin06g01180  1.824923e-95  94.738755   \n",
      "Bcin15g00270  3.840947e-93  92.415562   \n",
      "Bcin06g05330  1.424904e-91  90.846214   \n",
      "Bcin02g08770  1.408696e-87  86.851183   \n",
      "Bcin08g07120  3.709094e-80  79.430732   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin06g01180  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin15g00270  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin06g05330  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin02g08770  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin08g07120  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "(140, 11734)\n",
      "(140, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.05 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 27.13 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.27 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 24.75 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.27 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.10 seconds.\n",
      "\n",
      "Replacing 1 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.32 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_target amino acid  vs DMSO\n",
      "                  baseMean  log2FoldChange     lfcSE      stat        pvalue  \\\n",
      "Bcin01g00010     22.561056       -1.635648  0.319290 -5.122762  3.010921e-07   \n",
      "Bcin01g00020     15.925217       -1.323806  0.283749 -4.665413  3.079982e-06   \n",
      "Bcin01g00030     88.985519       -1.663859  0.318008 -5.232131  1.675666e-07   \n",
      "Bcin01g00040     30.744354       -1.509181  0.323829 -4.660428  3.155531e-06   \n",
      "Bcin01g00050     16.714336       -0.906204  0.215457 -4.205965  2.599708e-05   \n",
      "...                    ...             ...       ...       ...           ...   \n",
      "ENSRNA049513861   0.038673        0.404892  4.455341  0.090878  9.275896e-01   \n",
      "ENSRNA049513883   0.812979        0.122673  0.509416  0.240811  8.097014e-01   \n",
      "ENSRNA049514038   0.227493        0.099900  0.815930  0.122437  9.025531e-01   \n",
      "ENSRNA049514141   0.212830        0.526485  1.195216  0.440494  6.595794e-01   \n",
      "ENSRNA049514255   0.132210       -0.093975  1.174636 -0.080004  9.362342e-01   \n",
      "\n",
      "                     padj  \n",
      "Bcin01g00010     0.000003  \n",
      "Bcin01g00020     0.000022  \n",
      "Bcin01g00030     0.000002  \n",
      "Bcin01g00040     0.000023  \n",
      "Bcin01g00050     0.000153  \n",
      "...                   ...  \n",
      "ENSRNA049513861       NaN  \n",
      "ENSRNA049513883  0.879325  \n",
      "ENSRNA049514038       NaN  \n",
      "ENSRNA049514141       NaN  \n",
      "ENSRNA049514255       NaN  \n",
      "\n",
      "[11734 rows x 6 columns]\n",
      "                baseMean  log2FoldChange     lfcSE       stat        pvalue  \\\n",
      "Bcin03g09110   62.978836        3.656445  0.222851  16.407566  1.688397e-60   \n",
      "Bcin02g02050  127.939651        4.207716  0.267163  15.749606  6.911038e-56   \n",
      "Bcin02g04830  331.373932        3.517586  0.226219  15.549461  1.604502e-54   \n",
      "Bcin03g04660  294.713898        2.772497  0.184025  15.065879  2.715321e-51   \n",
      "Bcin09g01680  219.381424        3.022174  0.204752  14.760150  2.646899e-49   \n",
      "\n",
      "                      padj    -Log10P  \\\n",
      "Bcin03g09110  1.750699e-56  55.756788   \n",
      "Bcin02g02050  3.583028e-52  51.445750   \n",
      "Bcin02g04830  5.545693e-51  50.256044   \n",
      "Bcin03g04660  7.038790e-48  47.152502   \n",
      "Bcin09g01680  5.489138e-46  45.260496   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin03g09110  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin02g02050  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin02g04830  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin03g04660  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin09g01680  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "(90, 11711)\n",
      "(90, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 21.57 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.32 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 19.19 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.37 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.07 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.22 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_target ATP proton pump  vs DMSO\n",
      "                   baseMean  log2FoldChange     lfcSE       stat  \\\n",
      "Bcin01g00010      99.664558        3.030281  0.296825  10.208990   \n",
      "Bcin01g00020      72.789970        3.034250  0.303237  10.006197   \n",
      "Bcin01g00030     364.085175        2.865329  0.275572  10.397735   \n",
      "Bcin01g00040     143.299606        3.250796  0.309299  10.510210   \n",
      "Bcin01g00050      51.246094        2.701767  0.265073  10.192542   \n",
      "...                     ...             ...       ...        ...   \n",
      "ENSRNA049513883    0.507166        0.621095  0.596956   1.040437   \n",
      "ENSRNA049514038    0.465488        1.319338  0.769382   1.714802   \n",
      "ENSRNA049514141    0.198266        0.428148  2.212733   0.193493   \n",
      "ENSRNA049514212    0.006145        0.607783  4.307754   0.141090   \n",
      "ENSRNA049514255    0.170313       -0.389977  1.425904  -0.273495   \n",
      "\n",
      "                       pvalue          padj  \n",
      "Bcin01g00010     1.807371e-24  5.259451e-22  \n",
      "Bcin01g00020     1.431496e-23  3.631595e-21  \n",
      "Bcin01g00030     2.538958e-25  8.662225e-23  \n",
      "Bcin01g00040     7.752085e-26  2.949966e-23  \n",
      "Bcin01g00050     2.140925e-24  6.052090e-22  \n",
      "...                       ...           ...  \n",
      "ENSRNA049513883  2.981368e-01           NaN  \n",
      "ENSRNA049514038  8.638156e-02           NaN  \n",
      "ENSRNA049514141  8.465732e-01           NaN  \n",
      "ENSRNA049514212  8.877986e-01           NaN  \n",
      "ENSRNA049514255  7.844731e-01           NaN  \n",
      "\n",
      "[11711 rows x 6 columns]\n",
      "                 baseMean  log2FoldChange     lfcSE       stat        pvalue  \\\n",
      "Bcin13g00710   304.804443        4.064757  0.278307  14.605288  2.598827e-48   \n",
      "Bcin05g02520    31.382219        2.774455  0.195406  14.198420  9.370184e-46   \n",
      "Bcin14g00850  2017.472900        2.391672  0.178624  13.389420  6.972315e-41   \n",
      "Bcin02g00016   412.935974        4.521082  0.337715  13.387271  7.176926e-41   \n",
      "Bcin06g00650    57.099377        6.146586  0.465709  13.198335  8.969555e-40   \n",
      "\n",
      "                      padj    -Log10P  \\\n",
      "Bcin13g00710  2.571279e-44  43.589851   \n",
      "Bcin05g02520  4.635430e-42  41.333910   \n",
      "Bcin14g00850  1.775213e-37  36.750750   \n",
      "Bcin02g00016  1.775213e-37  36.750750   \n",
      "Bcin06g00650  1.774896e-36  35.750827   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin13g00710  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin05g02520  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin14g00850  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin02g00016  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin06g00650  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "(90, 11716)\n",
      "(90, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 21.89 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.34 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 19.05 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.37 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.07 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.27 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_target GPI  vs DMSO\n",
      "                   baseMean  log2FoldChange     lfcSE       stat  \\\n",
      "Bcin01g00010      90.611389        3.215010  0.239797  13.407198   \n",
      "Bcin01g00020      67.490135        3.216112  0.256789  12.524351   \n",
      "Bcin01g00030     346.105133        3.047380  0.234657  12.986527   \n",
      "Bcin01g00040     133.283447        3.322814  0.235391  14.116156   \n",
      "Bcin01g00050      47.119049        2.836537  0.214422  13.228732   \n",
      "...                     ...             ...       ...        ...   \n",
      "ENSRNA049513861    0.067227        1.057824  4.351015   0.243121   \n",
      "ENSRNA049513883    0.477996        0.765847  0.619380   1.236474   \n",
      "ENSRNA049514038    0.292437        0.416070  1.065299   0.390566   \n",
      "ENSRNA049514141    0.167294        1.210366  1.963421   0.616458   \n",
      "ENSRNA049514255    0.424937        1.065427  1.032075   1.032316   \n",
      "\n",
      "                       pvalue          padj  \n",
      "Bcin01g00010     5.487340e-41  1.184818e-38  \n",
      "Bcin01g00020     5.493908e-36  7.854803e-34  \n",
      "Bcin01g00030     1.458995e-38  2.708099e-36  \n",
      "Bcin01g00040     3.020232e-45  8.636231e-43  \n",
      "Bcin01g00050     5.988917e-40  1.173384e-37  \n",
      "...                       ...           ...  \n",
      "ENSRNA049513861  8.079114e-01           NaN  \n",
      "ENSRNA049513883  2.162824e-01  3.510300e-01  \n",
      "ENSRNA049514038  6.961179e-01  7.986971e-01  \n",
      "ENSRNA049514141  5.375925e-01           NaN  \n",
      "ENSRNA049514255  3.019242e-01  4.468883e-01  \n",
      "\n",
      "[11716 rows x 6 columns]\n",
      "                baseMean  log2FoldChange     lfcSE       stat         pvalue  \\\n",
      "Bcin02g02790  809.774414       -0.308856  0.013305 -23.213935  3.293129e-119   \n",
      "Bcin13g00710  496.754822        5.110758  0.242991  21.032672   3.295538e-98   \n",
      "Bcin06g00650  155.494049        7.905572  0.391326  20.202017   9.397927e-91   \n",
      "Bcin14g02510  228.967255        3.858875  0.195133  19.775646   4.826056e-87   \n",
      "Bcin15g02550  659.067871        3.644808  0.184844  19.718325   1.501168e-86   \n",
      "\n",
      "                       padj     -Log10P  \\\n",
      "Bcin02g02790  3.484131e-115  114.457906   \n",
      "Bcin13g00710   1.743340e-94   93.758618   \n",
      "Bcin06g00650   3.314336e-87   86.479604   \n",
      "Bcin14g02510   1.276492e-83   82.893982   \n",
      "Bcin15g02550   3.176472e-83   82.498055   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin02g02790                                    Under Threshold  \n",
      "Bcin13g00710  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin06g00650  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin14g02510  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin15g02550  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "(125, 11731)\n",
      "(125, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.05 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 24.24 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.40 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 24.10 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.30 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.09 seconds.\n",
      "\n",
      "Replacing 28 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.07 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.06 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.18 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_target proteins  vs DMSO\n",
      "                   baseMean  log2FoldChange     lfcSE      stat        pvalue  \\\n",
      "Bcin01g00010     101.071632        2.272039  0.430732  5.274825  1.328826e-07   \n",
      "Bcin01g00020      70.453117        2.106042  0.433847  4.854339  1.207890e-06   \n",
      "Bcin01g00030     333.708435        1.876680  0.403403  4.652120  3.285397e-06   \n",
      "Bcin01g00040     119.651634        2.032773  0.420296  4.836527  1.321275e-06   \n",
      "Bcin01g00050      50.356564        1.856028  0.390778  4.749575  2.038446e-06   \n",
      "...                     ...             ...       ...       ...           ...   \n",
      "ENSRNA049513861    0.050460       -0.174770  4.301723 -0.040628  9.675925e-01   \n",
      "ENSRNA049513883    0.473560       -0.209671  0.503481 -0.416442  6.770867e-01   \n",
      "ENSRNA049514038    0.309974       -0.030248  0.668309 -0.045261  9.638994e-01   \n",
      "ENSRNA049514141    0.179082        0.036948  1.468898  0.025154  9.799323e-01   \n",
      "ENSRNA049514255    0.247826        0.222373  0.770078  0.288767  7.727597e-01   \n",
      "\n",
      "                     padj  \n",
      "Bcin01g00010     0.000020  \n",
      "Bcin01g00020     0.000108  \n",
      "Bcin01g00030     0.000206  \n",
      "Bcin01g00040     0.000115  \n",
      "Bcin01g00050     0.000154  \n",
      "...                   ...  \n",
      "ENSRNA049513861       NaN  \n",
      "ENSRNA049513883       NaN  \n",
      "ENSRNA049514038       NaN  \n",
      "ENSRNA049514141       NaN  \n",
      "ENSRNA049514255       NaN  \n",
      "\n",
      "[11731 rows x 6 columns]\n",
      "                baseMean  log2FoldChange     lfcSE      stat        pvalue  \\\n",
      "Bcin15g00270   22.450977        4.143814  0.428363  9.673602  3.903901e-22   \n",
      "Bcin09g06180  187.807510        0.226279  0.025034  9.038732  1.584999e-19   \n",
      "Bcin15g05630   15.719965        3.501448  0.415357  8.429965  3.457682e-17   \n",
      "Bcin08g03460   43.960777        2.841591  0.342269  8.302211  1.021912e-16   \n",
      "Bcin08g04600   19.724592        1.889556  0.259802  7.273058  3.514384e-13   \n",
      "\n",
      "                      padj    -Log10P  \\\n",
      "Bcin15g00270  3.869157e-18  17.412384   \n",
      "Bcin09g06180  7.854462e-16  15.104884   \n",
      "Bcin15g05630  1.142303e-13  12.942219   \n",
      "Bcin08g03460  2.532043e-13  12.596529   \n",
      "Bcin08g04600  6.966211e-10   9.157003   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin15g00270  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin09g06180                                    Under Threshold  \n",
      "Bcin15g05630  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin08g03460  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin08g04600  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "(100, 11725)\n",
      "(100, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.04 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 22.12 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.33 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 20.58 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.07 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.07 seconds.\n",
      "\n",
      "Replacing 3 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.32 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_target Multi vs DMSO\n",
      "                   baseMean  log2FoldChange     lfcSE      stat        pvalue  \\\n",
      "Bcin01g00010      59.532852        1.354684  0.244044  5.550976  2.840794e-08   \n",
      "Bcin01g00020      48.326462        1.652089  0.250378  6.598368  4.157077e-11   \n",
      "Bcin01g00030     248.008163        1.452298  0.240525  6.038038  1.559997e-09   \n",
      "Bcin01g00040      88.203705        1.587878  0.243337  6.525426  6.780867e-11   \n",
      "Bcin01g00050      37.954979        1.443099  0.220508  6.544421  5.972646e-11   \n",
      "...                     ...             ...       ...       ...           ...   \n",
      "ENSRNA049513861    0.066802        0.003194  4.203390  0.000760  9.993937e-01   \n",
      "ENSRNA049513883    0.567927        0.634072  0.455015  1.393519  1.634628e-01   \n",
      "ENSRNA049514038    0.239567       -0.409030  0.853227 -0.479392  6.316599e-01   \n",
      "ENSRNA049514141    0.136935       -0.191360  2.225112 -0.086000  9.314662e-01   \n",
      "ENSRNA049514255    0.280231        0.335306  0.858662  0.390498  6.961683e-01   \n",
      "\n",
      "                         padj  \n",
      "Bcin01g00010     5.274815e-07  \n",
      "Bcin01g00020     1.963437e-09  \n",
      "Bcin01g00030     4.310167e-08  \n",
      "Bcin01g00040     2.903164e-09  \n",
      "Bcin01g00050     2.644555e-09  \n",
      "...                       ...  \n",
      "ENSRNA049513861           NaN  \n",
      "ENSRNA049513883  2.736088e-01  \n",
      "ENSRNA049514038           NaN  \n",
      "ENSRNA049514141           NaN  \n",
      "ENSRNA049514255           NaN  \n",
      "\n",
      "[11725 rows x 6 columns]\n",
      "                baseMean  log2FoldChange     lfcSE       stat        pvalue  \\\n",
      "Bcin02g00470  140.231018       -0.509339  0.030901 -16.482784  4.878576e-61   \n",
      "Bcin06g00024  363.677582        7.209818  0.632997  11.389970  4.691512e-30   \n",
      "Bcin12g05940   43.065216        4.794282  0.439201  10.915923  9.674146e-28   \n",
      "Bcin10g00540  357.689056       -0.661010  0.061181 -10.804192  3.288399e-27   \n",
      "Bcin12g05950   79.238411        4.933662  0.459751  10.731167  7.267260e-27   \n",
      "\n",
      "                      padj    -Log10P  \\\n",
      "Bcin02g00470  5.054693e-57  56.296305   \n",
      "Bcin06g00024  2.430438e-26  25.614316   \n",
      "Bcin12g05940  3.341127e-24  23.476107   \n",
      "Bcin10g00540  8.517776e-24  23.069674   \n",
      "Bcin12g05950  1.505922e-23  22.822198   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin02g00470                                    Under Threshold  \n",
      "Bcin06g00024  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin12g05940  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin10g00540                                    Under Threshold  \n",
      "Bcin12g05950  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "(65, 11694)\n",
      "(65, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 18.25 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.27 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 16.59 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.87 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.05 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.22 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_treatments BCS-AK75639 vs DMSO\n",
      "                   baseMean  log2FoldChange     lfcSE      stat        pvalue  \\\n",
      "Bcin01g00010      50.238426       -2.242940  0.376401 -5.958914  2.539194e-09   \n",
      "Bcin01g00020      35.899536       -1.549285  0.428173 -3.618357  2.964787e-04   \n",
      "Bcin01g00030     194.596222       -2.069022  0.393723 -5.255022  1.480070e-07   \n",
      "Bcin01g00040      68.095604       -2.025985  0.404054 -5.014148  5.326891e-07   \n",
      "Bcin01g00050      28.065269       -1.127219  0.334306 -3.371816  7.467443e-04   \n",
      "...                     ...             ...       ...       ...           ...   \n",
      "ENSRNA049513861    0.120283       -0.092657  6.066298 -0.015274  9.878135e-01   \n",
      "ENSRNA049513883    0.586549       -0.910195  1.390516 -0.654573  5.127425e-01   \n",
      "ENSRNA049514038    0.378312       -1.545692  1.857643 -0.832071  4.053687e-01   \n",
      "ENSRNA049514141    0.217103       -0.378974  6.061322 -0.062523  9.501461e-01   \n",
      "ENSRNA049514255    0.361215        0.261978  1.584903  0.165296  8.687109e-01   \n",
      "\n",
      "                         padj  \n",
      "Bcin01g00010     6.762241e-08  \n",
      "Bcin01g00020     2.730405e-03  \n",
      "Bcin01g00030     3.010545e-06  \n",
      "Bcin01g00040     9.864296e-06  \n",
      "Bcin01g00050     6.257996e-03  \n",
      "...                       ...  \n",
      "ENSRNA049513861           NaN  \n",
      "ENSRNA049513883  7.484458e-01  \n",
      "ENSRNA049514038           NaN  \n",
      "ENSRNA049514141           NaN  \n",
      "ENSRNA049514255           NaN  \n",
      "\n",
      "[11694 rows x 6 columns]\n",
      "                baseMean  log2FoldChange     lfcSE       stat         pvalue  \\\n",
      "Bcin04g00040   67.569283        5.315535  0.180605  29.431851  2.149463e-190   \n",
      "Bcin08g03220  179.206741        5.262121  0.197863  26.594748  7.807328e-156   \n",
      "Bcin13g04360  216.363190        3.758656  0.175340  21.436449  6.110210e-102   \n",
      "Bcin06g01180  144.069687        5.012282  0.250851  19.981143   8.036322e-89   \n",
      "Bcin12g02050   78.796753        3.016307  0.154037  19.581766   2.212184e-85   \n",
      "\n",
      "                       padj     -Log10P  \\\n",
      "Bcin04g00040  2.221040e-186  185.653444   \n",
      "Bcin08g03220  4.033656e-152  151.394301   \n",
      "Bcin13g04360   2.104560e-98   97.676839   \n",
      "Bcin06g01180   2.075983e-85   84.682776   \n",
      "Bcin12g02050   4.571700e-82   81.339922   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin04g00040  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin08g03220  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin13g04360  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin06g01180  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin12g02050  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "(65, 11691)\n",
      "(65, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 18.51 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.26 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 16.70 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.88 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.05 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.25 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_treatments BCS-CZ69404 vs DMSO\n",
      "                   baseMean  log2FoldChange     lfcSE      stat        pvalue  \\\n",
      "Bcin01g00010      49.176617       -3.109085  0.424392 -7.325982  2.371561e-13   \n",
      "Bcin01g00020      34.967892       -3.040081  0.526623 -5.772784  7.797246e-09   \n",
      "Bcin01g00030     189.267227       -4.202197  0.432577 -9.714341  2.619391e-22   \n",
      "Bcin01g00040      66.523430       -3.767485  0.490191 -7.685752  1.521005e-14   \n",
      "Bcin01g00050      27.366066       -2.077472  0.393788 -5.275609  1.323159e-07   \n",
      "...                     ...             ...       ...       ...           ...   \n",
      "ENSRNA049513861    0.119086        0.081967  6.055696  0.013535  9.892006e-01   \n",
      "ENSRNA049513883    0.600375       -0.202538  1.206958 -0.167809  8.667337e-01   \n",
      "ENSRNA049514038    0.374821       -1.390065  1.915867 -0.725554  4.681124e-01   \n",
      "ENSRNA049514141    0.215119       -0.206459  6.057931 -0.034081  9.728127e-01   \n",
      "ENSRNA049514255    0.301358       -1.174563  2.129670 -0.551523  5.812750e-01   \n",
      "\n",
      "                         padj  \n",
      "Bcin01g00010     8.390617e-12  \n",
      "Bcin01g00020     1.588824e-07  \n",
      "Bcin01g00030     2.407969e-20  \n",
      "Bcin01g00040     6.260358e-13  \n",
      "Bcin01g00050     2.259430e-06  \n",
      "...                       ...  \n",
      "ENSRNA049513861           NaN  \n",
      "ENSRNA049513883  9.475371e-01  \n",
      "ENSRNA049514038           NaN  \n",
      "ENSRNA049514141           NaN  \n",
      "ENSRNA049514255           NaN  \n",
      "\n",
      "[11691 rows x 6 columns]\n",
      "                baseMean  log2FoldChange     lfcSE       stat         pvalue  \\\n",
      "Bcin08g03220  135.753571        4.800912  0.206751  23.220714  2.812731e-119   \n",
      "Bcin02g05250   15.152718        3.687065  0.161332  22.853917  1.336021e-115   \n",
      "Bcin06g01180  202.513412        5.653546  0.253397  22.310997  2.889815e-110   \n",
      "Bcin05g04150  189.946640        2.034939  0.101387  20.071080   1.321023e-89   \n",
      "Bcin05g03530    4.745698        5.378876  0.275128  19.550451   4.088911e-85   \n",
      "\n",
      "                       padj     -Log10P  \\\n",
      "Bcin08g03220  2.905832e-115  114.536729   \n",
      "Bcin02g05250  6.901215e-112  111.161074   \n",
      "Bcin06g01180  9.951561e-107  106.002109   \n",
      "Bcin05g04150   3.411871e-86   85.467007   \n",
      "Bcin05g03530   8.448507e-82   81.073220   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin08g03220  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin02g05250  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin06g01180  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin05g04150  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin05g03530  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "(65, 11689)\n",
      "(65, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 18.30 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.29 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 16.65 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.85 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.05 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.25 seconds.\n",
      "\n",
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/pandas/core/arraylike.py:396: RuntimeWarning: divide by zero encountered in log10\n",
      "  result = getattr(ufunc, method)(*inputs, **kwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_treatments BCS-DG83654 vs DMSO\n",
      "                   baseMean  log2FoldChange     lfcSE       stat  \\\n",
      "Bcin01g00010      49.344002       -3.367699  0.432389  -7.788594   \n",
      "Bcin01g00020      35.125645       -3.089112  0.525047  -5.883502   \n",
      "Bcin01g00030     190.001282       -4.175914  0.407749 -10.241385   \n",
      "Bcin01g00040      66.756523       -3.953671  0.505907  -7.815013   \n",
      "Bcin01g00050      27.437544       -2.153494  0.397270  -5.420734   \n",
      "...                     ...             ...       ...        ...   \n",
      "ENSRNA049513861    0.119562       -0.002911  6.057105  -0.000481   \n",
      "ENSRNA049513883    0.584964       -0.858598  1.269915  -0.676107   \n",
      "ENSRNA049514038    0.392812       -0.890349  1.681336  -0.529548   \n",
      "ENSRNA049514141    0.228483        0.382353  5.989124   0.063841   \n",
      "ENSRNA049514255    0.338303       -0.180180  1.635024  -0.110201   \n",
      "\n",
      "                       pvalue          padj  \n",
      "Bcin01g00010     6.775878e-15  1.882089e-13  \n",
      "Bcin01g00020     4.016753e-09  6.225672e-08  \n",
      "Bcin01g00030     1.293703e-24  7.802879e-23  \n",
      "Bcin01g00040     5.495723e-15  1.542749e-13  \n",
      "Bcin01g00050     5.935482e-08  7.584626e-07  \n",
      "...                       ...           ...  \n",
      "ENSRNA049513861  9.996166e-01           NaN  \n",
      "ENSRNA049513883  4.989728e-01  6.938029e-01  \n",
      "ENSRNA049514038  5.964250e-01  7.695925e-01  \n",
      "ENSRNA049514141  9.490967e-01           NaN  \n",
      "ENSRNA049514255  9.122503e-01  9.545754e-01  \n",
      "\n",
      "[11689 rows x 6 columns]\n",
      "                baseMean  log2FoldChange     lfcSE       stat         pvalue  \\\n",
      "Bcin13g03900   44.202541        8.232525  0.214520  38.376531   0.000000e+00   \n",
      "Bcin08g03220  312.637177        6.171112  0.184851  33.384290  2.317743e-244   \n",
      "Bcin04g00040  121.673759        6.173828  0.232883  26.510408  7.352741e-155   \n",
      "Bcin12g00520   10.471669        5.274691  0.210306  25.080980  8.021036e-139   \n",
      "Bcin13g04360  331.055054        4.686429  0.192426  24.354437  5.202609e-131   \n",
      "\n",
      "                       padj     -Log10P  \\\n",
      "Bcin13g03900   0.000000e+00         inf   \n",
      "Bcin08g03220  1.223189e-240  239.912507   \n",
      "Bcin04g00040  2.586939e-151  150.587214   \n",
      "Bcin12g00520  2.116551e-135  134.674371   \n",
      "Bcin13g04360  1.098271e-127  126.959291   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin13g03900  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin08g03220  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin04g00040  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin12g00520  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin13g04360  Up-regulated over threshold of P < 0.05 \\n and...  \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "posx and posy should be finite values\n",
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/matplotlib/transforms.py:360: RuntimeWarning: invalid value encountered in scalar subtract\n",
      "  return points[1, 1] - points[0, 1]\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(65, 11688)\n",
      "(65, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 18.29 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.29 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 16.66 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.88 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.05 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.21 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_treatments BCS-AA31476 vs DMSO\n",
      "                   baseMean  log2FoldChange     lfcSE      stat        pvalue  \\\n",
      "Bcin01g00010      51.064087       -1.925452  0.467364 -4.119810  3.791852e-05   \n",
      "Bcin01g00020      35.075336       -1.839030  0.479902 -3.832096  1.270561e-04   \n",
      "Bcin01g00030     192.586182       -2.356082  0.444269 -5.303281  1.137399e-07   \n",
      "Bcin01g00040      67.642944       -1.939674  0.468712 -4.138309  3.498755e-05   \n",
      "Bcin01g00050      27.392111       -1.535696  0.407380 -3.769691  1.634496e-04   \n",
      "...                     ...             ...       ...       ...           ...   \n",
      "ENSRNA049513861    0.118204        0.207479  6.056922  0.034255  9.726740e-01   \n",
      "ENSRNA049513883    0.619625        0.648729  0.917075  0.707389  4.793248e-01   \n",
      "ENSRNA049514038    0.372353       -1.264445  1.921486 -0.658056  5.105022e-01   \n",
      "ENSRNA049514141    0.213493       -0.080656  6.053800 -0.013323  9.893699e-01   \n",
      "ENSRNA049514255    0.326530        0.196265  1.674268  0.117224  9.066824e-01   \n",
      "\n",
      "                     padj  \n",
      "Bcin01g00010     0.000481  \n",
      "Bcin01g00020     0.001420  \n",
      "Bcin01g00030     0.000003  \n",
      "Bcin01g00040     0.000450  \n",
      "Bcin01g00050     0.001753  \n",
      "...                   ...  \n",
      "ENSRNA049513861       NaN  \n",
      "ENSRNA049513883       NaN  \n",
      "ENSRNA049514038       NaN  \n",
      "ENSRNA049514141       NaN  \n",
      "ENSRNA049514255       NaN  \n",
      "\n",
      "[11688 rows x 6 columns]\n",
      "                baseMean  log2FoldChange     lfcSE       stat         pvalue  \\\n",
      "Bcin06g07150   37.505001        7.458567  0.264376  28.211943  4.173435e-175   \n",
      "Bcin14g02510  174.837875        2.775643  0.129714  21.398185  1.389097e-101   \n",
      "Bcin05g00740   17.833853        5.215220  0.247182  21.098695   8.176102e-99   \n",
      "Bcin13g00060   47.251343        3.874726  0.190782  20.309658   1.056369e-91   \n",
      "Bcin02g08770   54.927795        4.639424  0.257329  18.029166   1.150140e-72   \n",
      "\n",
      "                       padj     -Log10P  \\\n",
      "Bcin06g07150  4.215587e-171  170.375142   \n",
      "Bcin14g02510   7.015635e-98   97.153933   \n",
      "Bcin05g00740   2.752894e-95   94.560211   \n",
      "Bcin13g00060   2.667595e-88   87.573880   \n",
      "Bcin02g08770   2.323512e-69   68.633855   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin06g07150  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin14g02510  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin05g00740  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin13g00060  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin02g08770  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "(65, 11689)\n",
      "(65, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 18.29 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.29 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 16.41 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.85 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.05 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.17 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_treatments BCS-AI81874 vs DMSO\n",
      "                   baseMean  log2FoldChange     lfcSE      stat        pvalue  \\\n",
      "Bcin01g00010      47.305325       -3.561433  0.525879 -6.772345  1.267115e-11   \n",
      "Bcin01g00020      33.212162       -3.246106  0.599679 -5.413075  6.195165e-08   \n",
      "Bcin01g00030     183.418854       -2.915262  0.462682 -6.300790  2.961328e-10   \n",
      "Bcin01g00040      65.181442       -2.449575  0.513786 -4.767692  1.863486e-06   \n",
      "Bcin01g00050      26.517929       -1.576800  0.409895 -3.846840  1.196511e-04   \n",
      "...                     ...             ...       ...       ...           ...   \n",
      "ENSRNA049513861    0.113554        1.033085  6.107289  0.169156  8.656739e-01   \n",
      "ENSRNA049513883    0.551105        0.101451  1.162962  0.087235  9.304846e-01   \n",
      "ENSRNA049514038    0.365434        0.100975  1.614897  0.062527  9.501432e-01   \n",
      "ENSRNA049514141    0.204780        0.823965  6.102291  0.135025  8.925917e-01   \n",
      "ENSRNA049514255    0.310883        0.765359  1.715112  0.446244  6.554207e-01   \n",
      "\n",
      "                         padj  \n",
      "Bcin01g00010     4.348184e-10  \n",
      "Bcin01g00020     1.264622e-06  \n",
      "Bcin01g00030     8.449601e-09  \n",
      "Bcin01g00040     3.016919e-05  \n",
      "Bcin01g00050     1.305043e-03  \n",
      "...                       ...  \n",
      "ENSRNA049513861           NaN  \n",
      "ENSRNA049513883  9.725739e-01  \n",
      "ENSRNA049514038           NaN  \n",
      "ENSRNA049514141           NaN  \n",
      "ENSRNA049514255           NaN  \n",
      "\n",
      "[11689 rows x 6 columns]\n",
      "                baseMean  log2FoldChange     lfcSE       stat         pvalue  \\\n",
      "Bcin06g01180  269.022247        6.130023  0.245963  24.922517  4.242172e-137   \n",
      "Bcin06g05330   46.663017        6.023451  0.251099  23.988327  3.681420e-127   \n",
      "Bcin02g08770   92.392616        5.802341  0.245745  23.611273  2.952205e-123   \n",
      "Bcin14g01260   35.700996        4.751301  0.233390  20.357815   3.958484e-92   \n",
      "Bcin14g01290  124.939255        4.605058  0.229279  20.084944   9.993622e-90   \n",
      "\n",
      "                       padj     -Log10P  \\\n",
      "Bcin06g01180  4.381739e-133  132.358353   \n",
      "Bcin06g05330  1.901269e-123  122.720956   \n",
      "Bcin02g08770  1.016444e-119  118.992916   \n",
      "Bcin14g01260   1.022180e-88   87.990473   \n",
      "Bcin14g01290   1.913745e-86   85.718116   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin06g01180  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin06g05330  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin02g08770  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin14g01260  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin14g01290  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "(65, 11687)\n",
      "(65, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 18.25 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.32 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 16.41 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.86 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.05 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.40 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_treatments BCS-AD77064 vs DMSO\n",
      "                   baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "Bcin01g00010      52.156540       -0.583559  0.383094 -1.523277  0.127689   \n",
      "Bcin01g00020      36.865421       -0.340527  0.436996 -0.779245  0.435836   \n",
      "Bcin01g00030     199.188889       -0.688582  0.395163 -1.742528  0.081416   \n",
      "Bcin01g00040      69.769783       -0.607940  0.410914 -1.479480  0.139012   \n",
      "Bcin01g00050      28.736700       -0.242483  0.346663 -0.699479  0.484253   \n",
      "...                     ...             ...       ...       ...       ...   \n",
      "ENSRNA049513861    0.119079        0.037050  6.060235  0.006114  0.995122   \n",
      "ENSRNA049513883    0.600106       -0.292354  1.060800 -0.275598  0.782857   \n",
      "ENSRNA049514038    0.375831       -1.440003  1.918129 -0.750733  0.452813   \n",
      "ENSRNA049514141    0.215329       -0.246537  6.052625 -0.040732  0.967509   \n",
      "ENSRNA049514255    0.333143       -0.631095  1.895225 -0.332992  0.739140   \n",
      "\n",
      "                     padj  \n",
      "Bcin01g00010     0.611362  \n",
      "Bcin01g00020     0.884544  \n",
      "Bcin01g00030     0.503968  \n",
      "Bcin01g00040     0.633932  \n",
      "Bcin01g00050     0.905282  \n",
      "...                   ...  \n",
      "ENSRNA049513861       NaN  \n",
      "ENSRNA049513883       NaN  \n",
      "ENSRNA049514038       NaN  \n",
      "ENSRNA049514141       NaN  \n",
      "ENSRNA049514255       NaN  \n",
      "\n",
      "[11687 rows x 6 columns]\n",
      "                baseMean  log2FoldChange     lfcSE       stat        pvalue  \\\n",
      "Bcin14g02510  134.643692        1.752623  0.138702  12.635885  1.338795e-36   \n",
      "Bcin08g01960   69.820229        1.569403  0.150731  10.411963  2.186661e-25   \n",
      "Bcin02g06340   52.187115        1.807447  0.183786   9.834539  7.993302e-23   \n",
      "Bcin02g02830  125.926193        2.272884  0.234332   9.699438  3.031633e-22   \n",
      "Bcin01g09540   52.168926        1.514799  0.157369   9.625799  6.222166e-22   \n",
      "\n",
      "                      padj    -Log10P  \\\n",
      "Bcin14g02510  1.352317e-32  31.868922   \n",
      "Bcin08g01960  1.104373e-21  20.956884   \n",
      "Bcin02g06340  2.691345e-19  18.570031   \n",
      "Bcin02g02830  7.655631e-19  18.116019   \n",
      "Bcin01g09540  1.257002e-18  17.900664   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin14g02510  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin08g01960  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin02g06340  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin02g02830  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin01g09540  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "(65, 11690)\n",
      "(65, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 18.37 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.32 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 16.45 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.86 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.05 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.24 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_treatments BCS-CX66720 vs DMSO\n",
      "                   baseMean  log2FoldChange     lfcSE       stat  \\\n",
      "Bcin01g00010      94.399063        3.296880  0.323962  10.176747   \n",
      "Bcin01g00020      58.180901        3.302034  0.371923   8.878284   \n",
      "Bcin01g00030     302.581116        3.059050  0.356760   8.574540   \n",
      "Bcin01g00040     104.219513        3.094244  0.363115   8.521385   \n",
      "Bcin01g00050      39.696167        2.687734  0.292704   9.182422   \n",
      "...                     ...             ...       ...        ...   \n",
      "ENSRNA049513861    0.118491        0.123154  6.061476   0.020318   \n",
      "ENSRNA049513883    0.581623       -0.703399  1.306136  -0.538534   \n",
      "ENSRNA049514038    0.425594       -0.305899  1.493992  -0.204753   \n",
      "ENSRNA049514141    0.214231       -0.160565  6.062758  -0.026484   \n",
      "ENSRNA049514255    0.317102       -0.063571  1.689185  -0.037634   \n",
      "\n",
      "                       pvalue          padj  \n",
      "Bcin01g00010     2.518389e-24  1.913394e-22  \n",
      "Bcin01g00020     6.789968e-19  3.439206e-17  \n",
      "Bcin01g00030     9.948341e-18  4.507328e-16  \n",
      "Bcin01g00040     1.576571e-17  6.890172e-16  \n",
      "Bcin01g00050     4.215011e-20  2.392624e-18  \n",
      "...                       ...           ...  \n",
      "ENSRNA049513861  9.837901e-01           NaN  \n",
      "ENSRNA049513883  5.902082e-01           NaN  \n",
      "ENSRNA049514038  8.377653e-01           NaN  \n",
      "ENSRNA049514141  9.788715e-01           NaN  \n",
      "ENSRNA049514255  9.699795e-01           NaN  \n",
      "\n",
      "[11690 rows x 6 columns]\n",
      "                 baseMean  log2FoldChange     lfcSE       stat         pvalue  \\\n",
      "Bcin14g02510   429.024048        4.800436  0.137136  35.005041  1.885604e-268   \n",
      "Bcin15g03060  2248.085205       -5.833638  0.246338 -23.681459  5.598762e-124   \n",
      "Bcin08g04610   273.690948        3.666157  0.166611  22.004228  2.623500e-107   \n",
      "Bcin14g00610   147.226074        6.309824  0.289831  21.770665  4.401988e-105   \n",
      "Bcin06g07120    31.223885        3.645296  0.175243  20.801347   4.208201e-96   \n",
      "\n",
      "                       padj     -Log10P  \\\n",
      "Bcin14g02510  1.862411e-264  263.729924   \n",
      "Bcin15g03060  2.764949e-120  119.558313   \n",
      "Bcin08g04610  8.637437e-104  103.063615   \n",
      "Bcin14g00610  1.086961e-101  100.963786   \n",
      "Bcin06g07120   8.312880e-93   92.080248   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin14g02510  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin15g03060  Down-regulated over threshold of P < 0.05 \\n a...  \n",
      "Bcin08g04610  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin14g00610  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin06g07120  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "(70, 11704)\n",
      "(70, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 18.62 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.30 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 17.18 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.05 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.05 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.12 seconds.\n",
      "\n",
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/pandas/core/arraylike.py:396: RuntimeWarning: divide by zero encountered in log10\n",
      "  result = getattr(ufunc, method)(*inputs, **kwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_treatments BCS-CU93685 vs DMSO\n",
      "                   baseMean  log2FoldChange     lfcSE      stat        pvalue  \\\n",
      "Bcin01g00010      41.652573       -2.171865  0.335434 -6.474797  9.493929e-11   \n",
      "Bcin01g00020      35.234482        0.739029  0.344099  2.147724  3.173568e-02   \n",
      "Bcin01g00030     159.004623       -2.595106  0.356356 -7.282343  3.280720e-13   \n",
      "Bcin01g00040      55.769630       -2.527636  0.371344 -6.806727  9.984422e-12   \n",
      "Bcin01g00050      23.201717       -1.514430  0.341370 -4.436327  9.150689e-06   \n",
      "...                     ...             ...       ...       ...           ...   \n",
      "ENSRNA049513861    0.098036        1.409709  4.856148  0.290294  7.715916e-01   \n",
      "ENSRNA049513883    0.889162        2.465560  0.718996  3.429172  6.054259e-04   \n",
      "ENSRNA049514038    0.348136        0.438006  1.361067  0.321811  7.475962e-01   \n",
      "ENSRNA049514141    0.177129        1.156163  4.850098  0.238379  8.115870e-01   \n",
      "ENSRNA049514255    0.892011        3.559451  1.099153  3.238358  1.202197e-03   \n",
      "\n",
      "                         padj  \n",
      "Bcin01g00010     2.643753e-10  \n",
      "Bcin01g00020     4.618682e-02  \n",
      "Bcin01g00030     1.038614e-12  \n",
      "Bcin01g00040     2.935385e-11  \n",
      "Bcin01g00050     1.851974e-05  \n",
      "...                       ...  \n",
      "ENSRNA049513861  8.037769e-01  \n",
      "ENSRNA049513883  1.060129e-03  \n",
      "ENSRNA049514038  7.916448e-01  \n",
      "ENSRNA049514141  8.352074e-01  \n",
      "ENSRNA049514255  2.054689e-03  \n",
      "\n",
      "[11704 rows x 6 columns]\n",
      "                 baseMean  log2FoldChange     lfcSE       stat  pvalue  padj  \\\n",
      "Bcin01g07340  7504.262695        8.313461  0.172012  48.330596     0.0   0.0   \n",
      "Bcin08g00810   233.291275        3.584371  0.073217  48.955241     0.0   0.0   \n",
      "Bcin05g05440   111.313698        6.159994  0.146509  42.045238     0.0   0.0   \n",
      "Bcin13g04800  1000.342102        4.808774  0.115884  41.496432     0.0   0.0   \n",
      "Bcin15g01410   152.741699        4.985506  0.123176  40.474582     0.0   0.0   \n",
      "\n",
      "              -Log10P                                           Type hit  \n",
      "Bcin01g07340      inf  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin08g00810      inf  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin05g05440      inf  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin13g04800      inf  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin15g01410      inf  Up-regulated over threshold of P < 0.05 \\n and...  \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/matplotlib/transforms.py:360: RuntimeWarning: invalid value encountered in scalar subtract\n",
      "  return points[1, 1] - points[0, 1]\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n",
      "posx and posy should be finite values\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(65, 11691)\n",
      "(65, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 18.39 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.33 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 16.53 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.01 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.05 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.19 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_treatments BCS-DA15669 vs DMSO\n",
      "                   baseMean  log2FoldChange     lfcSE      stat        pvalue  \\\n",
      "Bcin01g00010      43.377655       -2.763178  0.597368 -4.625588  3.735373e-06   \n",
      "Bcin01g00020      35.028904        1.047080  0.429108  2.440133  1.468184e-02   \n",
      "Bcin01g00030     167.518219       -3.416897  0.517156 -6.607091  3.919459e-11   \n",
      "Bcin01g00040      59.317093       -2.940901  0.662184 -4.441213  8.945304e-06   \n",
      "Bcin01g00050      24.674110       -0.673236  0.480842 -1.400118  1.614779e-01   \n",
      "...                     ...             ...       ...       ...           ...   \n",
      "ENSRNA049513861    0.104820        2.516481  6.102861  0.412345  6.800869e-01   \n",
      "ENSRNA049513883    0.586769        2.164817  1.255418  1.724380  8.463933e-02   \n",
      "ENSRNA049514038    0.568665        1.650167  1.697677  0.972015  3.310433e-01   \n",
      "ENSRNA049514141    0.189420        2.290154  6.079601  0.376695  7.064004e-01   \n",
      "ENSRNA049514255    0.725976        4.093730  1.322659  3.095076  1.967626e-03   \n",
      "\n",
      "                         padj  \n",
      "Bcin01g00010     1.035127e-05  \n",
      "Bcin01g00020     2.557827e-02  \n",
      "Bcin01g00030     1.624003e-10  \n",
      "Bcin01g00040     2.387043e-05  \n",
      "Bcin01g00050     2.233264e-01  \n",
      "...                       ...  \n",
      "ENSRNA049513861  7.322014e-01  \n",
      "ENSRNA049513883  1.261563e-01  \n",
      "ENSRNA049514038  4.134178e-01  \n",
      "ENSRNA049514141  7.535002e-01  \n",
      "ENSRNA049514255  3.941917e-03  \n",
      "\n",
      "[11691 rows x 6 columns]\n",
      "                baseMean  log2FoldChange     lfcSE       stat         pvalue  \\\n",
      "Bcin08g03450  206.638382        5.693532  0.157738  36.094973  2.719617e-285   \n",
      "Bcin13g03020  154.663895        5.462271  0.154407  35.375880  4.011907e-274   \n",
      "Bcin08g05050  153.788040        5.369316  0.153215  35.044347  4.754288e-269   \n",
      "Bcin09g00650  339.291992        5.883479  0.170022  34.604259  2.180023e-262   \n",
      "Bcin15g05310   95.077156        5.344878  0.155450  34.383286  4.482421e-259   \n",
      "\n",
      "                       padj     -Log10P  \\\n",
      "Bcin08g03450  3.056034e-281  280.514842   \n",
      "Bcin13g03020  2.254090e-270  269.647029   \n",
      "Bcin08g05050  1.780798e-265  264.749385   \n",
      "Bcin09g00650  6.124231e-259  258.212948   \n",
      "Bcin15g05310  1.007379e-255  254.996807   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin08g03450  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin13g03020  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin08g05050  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin09g00650  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin15g05310  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "(65, 11696)\n",
      "(65, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 18.29 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.26 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 16.73 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.96 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.05 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.23 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_treatments BCS-AU85103 vs DMSO\n",
      "                   baseMean  log2FoldChange     lfcSE      stat        pvalue  \\\n",
      "Bcin01g00010      50.532028       -1.842177  0.400101 -4.604282  4.138913e-06   \n",
      "Bcin01g00020      35.319958       -2.002517  0.475173 -4.214291  2.505644e-05   \n",
      "Bcin01g00030     193.389191       -1.804058  0.380469 -4.741669  2.119652e-06   \n",
      "Bcin01g00040      67.255539       -2.280073  0.439494 -5.187948  2.126239e-07   \n",
      "Bcin01g00050      27.868528       -0.984964  0.340621 -2.891674  3.831951e-03   \n",
      "...                     ...             ...       ...       ...           ...   \n",
      "ENSRNA049513861    0.119032        0.040308  6.164412  0.006539  9.947828e-01   \n",
      "ENSRNA049513883    0.622146        1.083337  0.649345  1.668353  9.524566e-02   \n",
      "ENSRNA049514038    0.393972       -0.361843  1.516789 -0.238558  8.114480e-01   \n",
      "ENSRNA049514141    0.215223       -0.270537  6.127664 -0.044150  9.647847e-01   \n",
      "ENSRNA049514255    0.322814        0.519030  1.582536  0.327973  7.429318e-01   \n",
      "\n",
      "                     padj  \n",
      "Bcin01g00010     0.000145  \n",
      "Bcin01g00020     0.000632  \n",
      "Bcin01g00030     0.000087  \n",
      "Bcin01g00040     0.000012  \n",
      "Bcin01g00050     0.028492  \n",
      "...                   ...  \n",
      "ENSRNA049513861       NaN  \n",
      "ENSRNA049513883       NaN  \n",
      "ENSRNA049514038       NaN  \n",
      "ENSRNA049514141       NaN  \n",
      "ENSRNA049514255       NaN  \n",
      "\n",
      "[11696 rows x 6 columns]\n",
      "                baseMean  log2FoldChange     lfcSE       stat        pvalue  \\\n",
      "Bcin11g04530   26.520742        8.191487  0.428443  19.119210  1.747340e-81   \n",
      "Bcin14g02510  157.867142        2.429391  0.142493  17.049160  3.545921e-65   \n",
      "Bcin01g08740  237.350616        0.772975  0.068062  11.356854  6.857003e-30   \n",
      "Bcin11g04520    8.901452        1.709254  0.165522  10.326448  5.350542e-25   \n",
      "Bcin05g01060  311.419739       -1.451405  0.148436  -9.777962  1.400005e-22   \n",
      "\n",
      "                      padj    -Log10P  \\\n",
      "Bcin11g04530  1.647392e-77  76.783203   \n",
      "Bcin14g02510  1.671547e-61  60.776881   \n",
      "Bcin01g08740  2.154927e-26  25.666567   \n",
      "Bcin11g04520  1.261123e-21  20.899243   \n",
      "Bcin05g01060  2.639850e-19  18.578421   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin11g04530  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin14g02510  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin01g08740                                    Under Threshold  \n",
      "Bcin11g04520  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin05g01060  Down-regulated over threshold of P < 0.05 \\n a...  \n",
      "(65, 11688)\n",
      "(65, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 18.24 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.29 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 16.56 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.03 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.05 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 1.23 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Metadata_treatments BCS-DG52208 vs DMSO\n",
      "                   baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "Bcin01g00010      55.615284        0.215721  0.338990  0.636364  0.524539   \n",
      "Bcin01g00020      38.664478        0.284204  0.398121  0.713863  0.475312   \n",
      "Bcin01g00030     211.007950        0.230497  0.352927  0.653102  0.513691   \n",
      "Bcin01g00040      75.002197        0.488184  0.371394  1.314464  0.188690   \n",
      "Bcin01g00050      30.269388        0.446506  0.308964  1.445171  0.148410   \n",
      "...                     ...             ...       ...       ...       ...   \n",
      "ENSRNA049513861    0.121117       -0.284311  6.053867 -0.046964  0.962542   \n",
      "ENSRNA049513883    0.670726        0.792050  0.944137  0.838914  0.401517   \n",
      "ENSRNA049514038    0.394430       -1.192599  1.712483 -0.696415  0.486169   \n",
      "ENSRNA049514141    0.218949       -0.568261  6.050495 -0.093920  0.925173   \n",
      "ENSRNA049514255    0.362476       -0.031517  1.626626 -0.019376  0.984541   \n",
      "\n",
      "                     padj  \n",
      "Bcin01g00010     0.679410  \n",
      "Bcin01g00020     0.638339  \n",
      "Bcin01g00030     0.670697  \n",
      "Bcin01g00040     0.343468  \n",
      "Bcin01g00050     0.291825  \n",
      "...                   ...  \n",
      "ENSRNA049513861       NaN  \n",
      "ENSRNA049513883       NaN  \n",
      "ENSRNA049514038       NaN  \n",
      "ENSRNA049514141       NaN  \n",
      "ENSRNA049514255       NaN  \n",
      "\n",
      "[11688 rows x 6 columns]\n",
      "                baseMean  log2FoldChange     lfcSE       stat         pvalue  \\\n",
      "Bcin14g02510  295.126526        4.017403  0.138588  28.988076  9.300950e-185   \n",
      "Bcin14g02750  107.021774        5.197328  0.225194  23.079310  7.472058e-118   \n",
      "Bcin05g01000  115.101204        4.643046  0.236206  19.656746   5.061677e-86   \n",
      "Bcin16g02980  169.398270        2.061757  0.113187  18.215483   3.889697e-74   \n",
      "Bcin12g05690  126.425240        3.453461  0.190495  18.128912   1.884501e-73   \n",
      "\n",
      "                       padj     -Log10P  \\\n",
      "Bcin14g02510  9.184688e-181  180.036936   \n",
      "Bcin14g02750  3.689328e-114  113.433053   \n",
      "Bcin05g01000   1.666135e-82   81.778290   \n",
      "Bcin16g02980   9.602688e-71   70.017607   \n",
      "Bcin12g05690   3.721889e-70   69.429237   \n",
      "\n",
      "                                                       Type hit  \n",
      "Bcin14g02510  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin14g02750  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin05g01000  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin16g02980  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "Bcin12g05690  Up-regulated over threshold of P < 0.05 \\n and...  \n",
      "(65, 11689)\n",
      "(65, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/miniconda3/envs/bridge_env/lib/python3.11/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[46], line 6\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[38;5;28;01mcontinue\u001b[39;00m\n\u001b[1;32m      5\u001b[0m df_small \u001b[38;5;241m=\u001b[39m df[df[group]\u001b[38;5;241m.\u001b[39misin([treatment, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mDMSO\u001b[39m\u001b[38;5;124m\"\u001b[39m])]\u001b[38;5;241m.\u001b[39mcopy()\n\u001b[0;32m----> 6\u001b[0m summary_df \u001b[38;5;241m=\u001b[39m deseq(df_small, group, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mDMSO\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      7\u001b[0m make_volcano(summary_df, treatment)\n",
      "Cell \u001b[0;32mIn[41], line 72\u001b[0m, in \u001b[0;36mdeseq\u001b[0;34m(df, group, reference)\u001b[0m\n\u001b[1;32m     70\u001b[0m \u001b[38;5;66;03m# Fit size factors and dispersions\u001b[39;00m\n\u001b[1;32m     71\u001b[0m dds\u001b[38;5;241m.\u001b[39mfit_size_factors()\n\u001b[0;32m---> 72\u001b[0m dds\u001b[38;5;241m.\u001b[39mfit_genewise_dispersions()\n\u001b[1;32m     73\u001b[0m dds\u001b[38;5;241m.\u001b[39mfit_dispersion_trend()\n\u001b[1;32m     74\u001b[0m dds\u001b[38;5;241m.\u001b[39mfit_dispersion_prior()\n",
      "File \u001b[0;32m~/miniconda3/envs/bridge_env/lib/python3.11/site-packages/pydeseq2/dds.py:762\u001b[0m, in \u001b[0;36mDeseqDataSet.fit_genewise_dispersions\u001b[0;34m(self, vst)\u001b[0m\n\u001b[1;32m    760\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFitting dispersions...\u001b[39m\u001b[38;5;124m\"\u001b[39m, file\u001b[38;5;241m=\u001b[39msys\u001b[38;5;241m.\u001b[39mstderr)\n\u001b[1;32m    761\u001b[0m start \u001b[38;5;241m=\u001b[39m time\u001b[38;5;241m.\u001b[39mtime()\n\u001b[0;32m--> 762\u001b[0m dispersions_, l_bfgs_b_converged_ \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39minference\u001b[38;5;241m.\u001b[39malpha_mle(\n\u001b[1;32m    763\u001b[0m     counts\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mX[:, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mnon_zero_idx],\n\u001b[1;32m    764\u001b[0m     design_matrix\u001b[38;5;241m=\u001b[39mdesign_matrix,\n\u001b[1;32m    765\u001b[0m     mu\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlayers[mu_param_name][:, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mnon_zero_idx],\n\u001b[1;32m    766\u001b[0m     alpha_hat\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvarm[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_MoM_dispersions\u001b[39m\u001b[38;5;124m\"\u001b[39m][\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mnon_zero_idx],\n\u001b[1;32m    767\u001b[0m     min_disp\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmin_disp,\n\u001b[1;32m    768\u001b[0m     max_disp\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmax_disp,\n\u001b[1;32m    769\u001b[0m )\n\u001b[1;32m    770\u001b[0m end \u001b[38;5;241m=\u001b[39m time\u001b[38;5;241m.\u001b[39mtime()\n\u001b[1;32m    772\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mquiet:\n",
      "File \u001b[0;32m~/miniconda3/envs/bridge_env/lib/python3.11/site-packages/pydeseq2/default_inference.py:140\u001b[0m, in \u001b[0;36mDefaultInference.alpha_mle\u001b[0;34m(self, counts, design_matrix, mu, alpha_hat, min_disp, max_disp, prior_disp_var, cr_reg, prior_reg, optimizer)\u001b[0m\n\u001b[1;32m    126\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21malpha_mle\u001b[39m(  \u001b[38;5;66;03m# noqa: D102\u001b[39;00m\n\u001b[1;32m    127\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    128\u001b[0m     counts: np\u001b[38;5;241m.\u001b[39mndarray,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    137\u001b[0m     optimizer: Literal[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mBFGS\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mL-BFGS-B\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mL-BFGS-B\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    138\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28mtuple\u001b[39m[np\u001b[38;5;241m.\u001b[39mndarray, np\u001b[38;5;241m.\u001b[39mndarray]:\n\u001b[1;32m    139\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m parallel_backend(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backend, inner_max_num_threads\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m):\n\u001b[0;32m--> 140\u001b[0m         res \u001b[38;5;241m=\u001b[39m Parallel(\n\u001b[1;32m    141\u001b[0m             n_jobs\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mn_cpus,\n\u001b[1;32m    142\u001b[0m             verbose\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_joblib_verbosity,\n\u001b[1;32m    143\u001b[0m             batch_size\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_batch_size,\n\u001b[1;32m    144\u001b[0m         )(\n\u001b[1;32m    145\u001b[0m             delayed(utils\u001b[38;5;241m.\u001b[39mfit_alpha_mle)(\n\u001b[1;32m    146\u001b[0m                 counts\u001b[38;5;241m=\u001b[39mcounts[:, i],\n\u001b[1;32m    147\u001b[0m                 design_matrix\u001b[38;5;241m=\u001b[39mdesign_matrix,\n\u001b[1;32m    148\u001b[0m                 mu\u001b[38;5;241m=\u001b[39mmu[:, i],\n\u001b[1;32m    149\u001b[0m                 alpha_hat\u001b[38;5;241m=\u001b[39malpha_hat[i],\n\u001b[1;32m    150\u001b[0m                 min_disp\u001b[38;5;241m=\u001b[39mmin_disp,\n\u001b[1;32m    151\u001b[0m                 max_disp\u001b[38;5;241m=\u001b[39mmax_disp,\n\u001b[1;32m    152\u001b[0m                 prior_disp_var\u001b[38;5;241m=\u001b[39mprior_disp_var,\n\u001b[1;32m    153\u001b[0m                 cr_reg\u001b[38;5;241m=\u001b[39mcr_reg,\n\u001b[1;32m    154\u001b[0m                 prior_reg\u001b[38;5;241m=\u001b[39mprior_reg,\n\u001b[1;32m    155\u001b[0m                 optimizer\u001b[38;5;241m=\u001b[39moptimizer,\n\u001b[1;32m    156\u001b[0m             )\n\u001b[1;32m    157\u001b[0m             \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(counts\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m1\u001b[39m])\n\u001b[1;32m    158\u001b[0m         )\n\u001b[1;32m    159\u001b[0m     res \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mzip\u001b[39m(\u001b[38;5;241m*\u001b[39mres)\n\u001b[1;32m    160\u001b[0m     dispersions_, l_bfgs_b_converged_ \u001b[38;5;241m=\u001b[39m (np\u001b[38;5;241m.\u001b[39marray(m) \u001b[38;5;28;01mfor\u001b[39;00m m \u001b[38;5;129;01min\u001b[39;00m res)\n",
      "File \u001b[0;32m~/miniconda3/envs/bridge_env/lib/python3.11/site-packages/joblib/parallel.py:1952\u001b[0m, in \u001b[0;36mParallel.__call__\u001b[0;34m(self, iterable)\u001b[0m\n\u001b[1;32m   1946\u001b[0m \u001b[38;5;66;03m# The first item from the output is blank, but it makes the interpreter\u001b[39;00m\n\u001b[1;32m   1947\u001b[0m \u001b[38;5;66;03m# progress until it enters the Try/Except block of the generator and\u001b[39;00m\n\u001b[1;32m   1948\u001b[0m \u001b[38;5;66;03m# reach the first `yield` statement. This starts the aynchronous\u001b[39;00m\n\u001b[1;32m   1949\u001b[0m \u001b[38;5;66;03m# dispatch of the tasks to the workers.\u001b[39;00m\n\u001b[1;32m   1950\u001b[0m \u001b[38;5;28mnext\u001b[39m(output)\n\u001b[0;32m-> 1952\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m output \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mreturn_generator \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mlist\u001b[39m(output)\n",
      "File \u001b[0;32m~/miniconda3/envs/bridge_env/lib/python3.11/site-packages/joblib/parallel.py:1595\u001b[0m, in \u001b[0;36mParallel._get_outputs\u001b[0;34m(self, iterator, pre_dispatch)\u001b[0m\n\u001b[1;32m   1592\u001b[0m     \u001b[38;5;28;01myield\u001b[39;00m\n\u001b[1;32m   1594\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backend\u001b[38;5;241m.\u001b[39mretrieval_context():\n\u001b[0;32m-> 1595\u001b[0m         \u001b[38;5;28;01myield from\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_retrieve()\n\u001b[1;32m   1597\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mGeneratorExit\u001b[39;00m:\n\u001b[1;32m   1598\u001b[0m     \u001b[38;5;66;03m# The generator has been garbage collected before being fully\u001b[39;00m\n\u001b[1;32m   1599\u001b[0m     \u001b[38;5;66;03m# consumed. This aborts the remaining tasks if possible and warn\u001b[39;00m\n\u001b[1;32m   1600\u001b[0m     \u001b[38;5;66;03m# the user if necessary.\u001b[39;00m\n\u001b[1;32m   1601\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_exception \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/bridge_env/lib/python3.11/site-packages/joblib/parallel.py:1707\u001b[0m, in \u001b[0;36mParallel._retrieve\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1702\u001b[0m \u001b[38;5;66;03m# If the next job is not ready for retrieval yet, we just wait for\u001b[39;00m\n\u001b[1;32m   1703\u001b[0m \u001b[38;5;66;03m# async callbacks to progress.\u001b[39;00m\n\u001b[1;32m   1704\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m ((\u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_jobs) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m\n\u001b[1;32m   1705\u001b[0m     (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_jobs[\u001b[38;5;241m0\u001b[39m]\u001b[38;5;241m.\u001b[39mget_status(\n\u001b[1;32m   1706\u001b[0m         timeout\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtimeout) \u001b[38;5;241m==\u001b[39m TASK_PENDING)):\n\u001b[0;32m-> 1707\u001b[0m     time\u001b[38;5;241m.\u001b[39msleep(\u001b[38;5;241m0.01\u001b[39m)\n\u001b[1;32m   1708\u001b[0m     \u001b[38;5;28;01mcontinue\u001b[39;00m\n\u001b[1;32m   1710\u001b[0m \u001b[38;5;66;03m# We need to be careful: the job list can be filling up as\u001b[39;00m\n\u001b[1;32m   1711\u001b[0m \u001b[38;5;66;03m# we empty it and Python list are not thread-safe by\u001b[39;00m\n\u001b[1;32m   1712\u001b[0m \u001b[38;5;66;03m# default hence the use of the lock\u001b[39;00m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "for group in [\"Metadata_target\", \"Metadata_treatments\", \"Metadata_leiden\"]:\n",
    "    for treatment in df[group].unique():\n",
    "        if treatment == \"DMSO\":\n",
    "            continue\n",
    "        df_small = df[df[group].isin([treatment, \"DMSO\"])].copy()\n",
    "        summary_df = deseq(df_small, group, \"DMSO\")\n",
    "        make_volcano(summary_df, treatment)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.lines import Line2D\n",
    "\n",
    "def volcano_comparison(treatments):\n",
    "    \"\"\"\n",
    "    Create an interactive volcano plot for multiple treatments and save as HTML. \n",
    "    Attach meta data to individual genes.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    treatments : list of strings\n",
    "        List of treatment names.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    fig : object\n",
    "        The figure object of the volcano plot.\n",
    "    \"\"\"\n",
    "    fig, ax = plt.subplots(figsize=(22, 12))\n",
    "    target, treatments = treatments\n",
    "    colour_dict = {\n",
    "        \"Under Threshold\": \"lightgrey\",\n",
    "        \"Up-regulated\": \"cornflowerblue\",\n",
    "        \"Down-regulated\": \"crimson\"\n",
    "    }\n",
    "\n",
    "    markers = ['o', 'X', 'D', '^', 'v', 'p', '*']\n",
    "    filtered_dataframes, overlapping_significant_genes = [], []\n",
    "    \n",
    "\n",
    "    for i, treatment in enumerate(treatments):\n",
    "        table_path = f\"{treatment}_differential_expression.csv\"\n",
    "        if not os.path.isfile(table_path):\n",
    "            print(f\"File {table_path} not found.\")\n",
    "            continue\n",
    "        \n",
    "        data_frame = pd.read_csv(table_path, index_col=0)\n",
    "        \n",
    "        # Check for duplicate indices\n",
    "        duplicates = data_frame.index.duplicated(keep=False)\n",
    "\n",
    "        # Create a new index with suffixes for duplicates\n",
    "        if duplicates.any():\n",
    "            # Create a dictionary to count occurrences of each index\n",
    "            count_dict = {}\n",
    "            new_index = []\n",
    "\n",
    "            for idx in data_frame.index:\n",
    "                if idx in count_dict:\n",
    "                    count_dict[idx] += 1\n",
    "                    new_index.append(f\"{idx}_{count_dict[idx]}\")\n",
    "                else:\n",
    "                    count_dict[idx] = 1\n",
    "                    new_index.append(idx)\n",
    "\n",
    "            # Assign the new unique index to the DataFrame\n",
    "            data_frame.index = new_index\n",
    "        \n",
    "        data_frame[\"-Log10P\"] = np.maximum(-np.log10(data_frame[\"padj\"]), 0)\n",
    "        data_frame[\"Type hit\"] = \"Under Threshold\"\n",
    "        data_frame.loc[(data_frame[\"log2FoldChange\"].abs() > 1) & (data_frame[\"padj\"] < 0.05), \"Type hit\"] = \"Up-regulated\"\n",
    "        data_frame.rename(columns={\n",
    "            \"log2FoldChange\": \"Log(2) Fold-Change\",\n",
    "            \"-Log10P\": \"-Log(10) of adjusted P-value\"\n",
    "        }, inplace=True)\n",
    "        \n",
    "        significant_data = data_frame[data_frame[\"Type hit\"] != \"Under Threshold\"].sort_values(\"-Log(10) of adjusted P-value\", ascending=False)\n",
    "        significant_data[\"Treatment\"] = treatment\n",
    "        filtered_dataframes.append(significant_data)\n",
    "        overlapping_significant_genes.append(set(significant_data.index))\n",
    "\n",
    "        sns.scatterplot(data=data_frame, x=\"Log(2) Fold-Change\", y=\"-Log(10) of adjusted P-value\", alpha=0.6, color=\"gainsboro\", \n",
    "                        edgecolor='none', marker=markers[i % len(markers)], ax=ax)\n",
    "    if len(overlapping_significant_genes) < 2:\n",
    "        print(\"No overlapping genes found.\")\n",
    "        return\n",
    "    intersection_of_genes = set.intersection(*overlapping_significant_genes)\n",
    "    n = min(len(intersection_of_genes), 10)\n",
    "    if n <= 1:\n",
    "        print(\"No overlapping genes found.\")\n",
    "        return\n",
    "    colors = plt.get_cmap(\"tab10\")(np.linspace(0, 1, n))\n",
    "    \n",
    "    overlapping_gene_data = pd.concat([df[df.index.isin(intersection_of_genes)] for df in filtered_dataframes], axis=0)\n",
    "\n",
    "    if not overlapping_gene_data.empty:\n",
    "        overlapping_gene_data[\"Top 10 overlapping\\ngenes by significance\"] = overlapping_gene_data.index\n",
    "        top_genes = overlapping_gene_data.groupby(\"Top 10 overlapping\\ngenes by significance\")[\"-Log(10) of adjusted P-value\"].max().nlargest(10).index\n",
    "        overlapping_gene_data = overlapping_gene_data[overlapping_gene_data[\"Top 10 overlapping\\ngenes by significance\"].isin(top_genes)]\n",
    "\n",
    "        # Sort the overlapping_gene_data by -Log(10) of adjusted P-value in descending order\n",
    "        overlapping_gene_data = overlapping_gene_data.sort_values(\"-Log(10) of adjusted P-value\", ascending=False)\n",
    "\n",
    "        # Create a sorted list of unique genes based on their order in the sorted DataFrame\n",
    "        sorted_genes = overlapping_gene_data[\"Top 10 overlapping\\ngenes by significance\"].unique()\n",
    "\n",
    "        for i, gene in enumerate(sorted_genes):\n",
    "            subset = overlapping_gene_data[overlapping_gene_data[\"Top 10 overlapping\\ngenes by significance\"] == gene]\n",
    "            ax.plot(subset[\"Log(2) Fold-Change\"], subset[\"-Log(10) of adjusted P-value\"], ls=\"--\", alpha=0.4, c=colors[i])\n",
    "        \n",
    "        #overlapping_gene_data[\"Top 10 overlapping\\ngenes by significance\"] = [gene_mapping[x].split(\",\")[0] for x in overlapping_gene_data[\"Top 10 overlapping\\ngenes by significance\"]]\n",
    "        overlapping_gene_data = overlapping_gene_data.drop_duplicates(subset=[\"Top 10 overlapping\\ngenes by significance\", \"Treatment\"], keep=\"first\")\n",
    "        sns.scatterplot(\n",
    "            data=overlapping_gene_data.reset_index(drop = True),\n",
    "            x=\"Log(2) Fold-Change\",\n",
    "            y=\"-Log(10) of adjusted P-value\",\n",
    "            hue=\"Top 10 overlapping\\ngenes by significance\", \n",
    "            style=\"Treatment\",\n",
    "            markers=markers[:len(intersection_of_genes)],\n",
    "            palette=colors, \n",
    "            s=100,\n",
    "            edgecolor=\"white\",\n",
    "            linewidth=1, ax=ax\n",
    "        )\n",
    "    # Add threshold lines\n",
    "    ax.axhline(y=-np.log10(0.05), lw=2, ls=\"--\", c=\"grey\")\n",
    "    ax.axvline(x=1, lw=2, ls=\"--\", c=\"grey\")\n",
    "    ax.axvline(x=-1, lw=2, ls=\"--\", c=\"grey\")\n",
    "    \n",
    "    ax.spines[['right', 'top']].set_visible(False)\n",
    "    ax.grid(True, alpha=0.2)\n",
    "    ax.tick_params(axis='both', which='major', labelsize=12)\n",
    "    sns.move_legend(ax, \"upper left\", bbox_to_anchor=(1, 1))\n",
    "\n",
    "    ax.set_title(f\"Volcano Plot of Target {target}\", fontsize=16)\n",
    "    plt.savefig(f\"volcano_plot_{target}.png\", bbox_inches='tight', dpi=300)\n",
    "    plt.savefig(f\"volcano_plot_{target}.svg\")\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "treatment_groupings = list(df.groupby([\"Metadata_target\"])[\"Metadata_treatments\"])\n",
    "treatment_groupings = [(x[0][0],list(x[1].unique())) for x in treatment_groupings if len(x[1]) > 1]\n",
    "treatment_groupings = [x for x in treatment_groupings if len(x[1]) > 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for grouping in treatment_groupings:\n",
    "    print(grouping)\n",
    "    volcano_comparison(grouping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bridge_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
